
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/pheval/developing_a_pheval_plugin/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-8.5.10">
    
    
      
        <title>Developing a PhEval Plugin - PhEval</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.975780f9.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.2505c338.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/mermaid@8.5.1/dist/mermaid.css">
    
      <link rel="stylesheet" href="../mermaid.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#developing-a-pheval-plugin" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="PhEval" class="md-header__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PhEval
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Developing a PhEval Plugin
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Night Mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Night Mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
            </label>
          
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Day Mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Day Mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="PhEval" class="md-nav__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    PhEval
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" >
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Pipeline
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pipeline" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pipeline
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../exomiser_pipeline/" class="md-nav__link">
        PhEval Pipeline Exomiser Runner
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline/" class="md-nav__link">
        PhEval Pipeline
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        About
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../contact/" class="md-nav__link">
        Contact Us
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          API Documentation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API Documentation" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          API Documentation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/cli/" class="md-nav__link">
        Cli
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/config_parser/" class="md-nav__link">
        Config parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/constants/" class="md-nav__link">
        Constants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/run_metadata/" class="md-nav__link">
        Run metadata
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_5" type="checkbox" id="__nav_5_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_5">
          analyse
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="analyse" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_5">
          <span class="md-nav__icon md-icon"></span>
          analyse
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/analysis/" class="md-nav__link">
        Analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/benchmark_generator/" class="md-nav__link">
        Benchmark generator
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/benchmarking_data/" class="md-nav__link">
        Benchmarking data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/disease_prioritisation_analysis/" class="md-nav__link">
        Disease prioritisation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/gene_prioritisation_analysis/" class="md-nav__link">
        Gene prioritisation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/generate_plots/" class="md-nav__link">
        Generate plots
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/generate_summary_outputs/" class="md-nav__link">
        Generate summary outputs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/parse_benchmark_summary/" class="md-nav__link">
        Parse benchmark summary
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/parse_pheval_result/" class="md-nav__link">
        Parse pheval result
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/prioritisation_rank_recorder/" class="md-nav__link">
        Prioritisation rank recorder
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/prioritisation_result_types/" class="md-nav__link">
        Prioritisation result types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/rank_stats/" class="md-nav__link">
        Rank stats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/run_data_parser/" class="md-nav__link">
        Run data parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/analyse/variant_prioritisation_analysis/" class="md-nav__link">
        Variant prioritisation analysis
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_6" type="checkbox" id="__nav_5_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_6">
          infra
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="infra" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_6">
          <span class="md-nav__icon md-icon"></span>
          infra
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/infra/exomiserdb/" class="md-nav__link">
        Exomiserdb
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_7" type="checkbox" id="__nav_5_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_7">
          post_processing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="post_processing" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_7">
          <span class="md-nav__icon md-icon"></span>
          post_processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/post_processing/post_processing/" class="md-nav__link">
        Post processing
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_8" type="checkbox" id="__nav_5_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_8">
          prepare
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="prepare" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_8">
          <span class="md-nav__icon md-icon"></span>
          prepare
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/prepare/create_noisy_phenopackets/" class="md-nav__link">
        Create noisy phenopackets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/prepare/create_spiked_vcf/" class="md-nav__link">
        Create spiked vcf
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/prepare/custom_exceptions/" class="md-nav__link">
        Custom exceptions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/prepare/update_phenopacket/" class="md-nav__link">
        Update phenopacket
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_9" type="checkbox" id="__nav_5_9" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_9">
          runners
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="runners" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_9">
          <span class="md-nav__icon md-icon"></span>
          runners
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/runners/runner/" class="md-nav__link">
        Runner
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_10" type="checkbox" id="__nav_5_10" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_10">
          utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="utils" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_10">
          <span class="md-nav__icon md-icon"></span>
          utils
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/utils/exomiser/" class="md-nav__link">
        Exomiser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/utils/file_utils/" class="md-nav__link">
        File utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/utils/phenopacket_utils/" class="md-nav__link">
        Phenopacket utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/utils/semsim_utils/" class="md-nav__link">
        Semsim utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../api/pheval/utils/utils/" class="md-nav__link">
        Utils
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Resources for contributors
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Resources for contributors" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Resources for contributors
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Developing a PhEval Plugin
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Developing a PhEval Plugin
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#description" class="md-nav__link">
    Description
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.runners.runner.PhEvalRunner" class="md-nav__link">
    src.pheval.runners.runner.PhEvalRunner
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-by-step-plugin-development-process" class="md-nav__link">
    Step-by-Step Plugin Development Process
  </a>
  
    <nav class="md-nav" aria-label="Step-by-Step Plugin Development Process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-sphintoxetry-cookiecutter-scaffold" class="md-nav__link">
    1. Sphintoxetry-cookiecutter scaffold
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-further-setup" class="md-nav__link">
    2. Further setup
  </a>
  
    <nav class="md-nav" aria-label="2. Further setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-poetry-if-you-havent-already" class="md-nav__link">
    Install poetry if you haven't already.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-dependencies" class="md-nav__link">
    Install dependencies
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-pheval-dependency" class="md-nav__link">
    Add PhEval dependency
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-tox-to-see-if-the-setup-works" class="md-nav__link">
    Run tox to see if the setup works
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-implement-pheval-custom-runner" class="md-nav__link">
    3. Implement PhEval Custom Runner
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-add-pheval-plugins-section-to-the-pyprojecttoml-file" class="md-nav__link">
    4. Add PhEval Plugins section to the pyproject.toml file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-implementing-pheval-helper-methods" class="md-nav__link">
    5. Implementing PhEval helper methods
  </a>
  
    <nav class="md-nav" aria-label="5. Implementing PhEval helper methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utility-methods" class="md-nav__link">
    Utility methods
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    src.pheval.utils.phenopacket_utils.PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="src.pheval.utils.phenopacket_utils.PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#additional-tool-specific-configurations" class="md-nav__link">
    Additional tool-specific configurations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#post-processing-methods" class="md-nav__link">
    Post-processing methods
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalGeneResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalGeneResult
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalVariantResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalVariantResult
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalDiseaseResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalDiseaseResult
  </a>
  
    <nav class="md-nav" aria-label="src.pheval.post_processing.post_processing.PhEvalDiseaseResult">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#adding-metadata-to-the-resultsyml" class="md-nav__link">
    Adding metadata to the results.yml
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-test-it" class="md-nav__link">
    6. Test it.
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        Contributions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../styleguide/" class="md-nav__link">
        Monarch Style Guide for PhEval
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../CODE_OF_CONDUCT/" class="md-nav__link">
        Contributor Covenant Code of Conduct
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../roadmap/" class="md-nav__link">
        Roadmap
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#description" class="md-nav__link">
    Description
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.runners.runner.PhEvalRunner" class="md-nav__link">
    src.pheval.runners.runner.PhEvalRunner
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-by-step-plugin-development-process" class="md-nav__link">
    Step-by-Step Plugin Development Process
  </a>
  
    <nav class="md-nav" aria-label="Step-by-Step Plugin Development Process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-sphintoxetry-cookiecutter-scaffold" class="md-nav__link">
    1. Sphintoxetry-cookiecutter scaffold
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-further-setup" class="md-nav__link">
    2. Further setup
  </a>
  
    <nav class="md-nav" aria-label="2. Further setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#install-poetry-if-you-havent-already" class="md-nav__link">
    Install poetry if you haven't already.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#install-dependencies" class="md-nav__link">
    Install dependencies
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-pheval-dependency" class="md-nav__link">
    Add PhEval dependency
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-tox-to-see-if-the-setup-works" class="md-nav__link">
    Run tox to see if the setup works
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-implement-pheval-custom-runner" class="md-nav__link">
    3. Implement PhEval Custom Runner
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-add-pheval-plugins-section-to-the-pyprojecttoml-file" class="md-nav__link">
    4. Add PhEval Plugins section to the pyproject.toml file
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-implementing-pheval-helper-methods" class="md-nav__link">
    5. Implementing PhEval helper methods
  </a>
  
    <nav class="md-nav" aria-label="5. Implementing PhEval helper methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#utility-methods" class="md-nav__link">
    Utility methods
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    src.pheval.utils.phenopacket_utils.PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="src.pheval.utils.phenopacket_utils.PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#additional-tool-specific-configurations" class="md-nav__link">
    Additional tool-specific configurations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#post-processing-methods" class="md-nav__link">
    Post-processing methods
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalGeneResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalGeneResult
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalVariantResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalVariantResult
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.post_processing.post_processing.PhEvalDiseaseResult" class="md-nav__link">
    src.pheval.post_processing.post_processing.PhEvalDiseaseResult
  </a>
  
    <nav class="md-nav" aria-label="src.pheval.post_processing.post_processing.PhEvalDiseaseResult">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#adding-metadata-to-the-resultsyml" class="md-nav__link">
    Adding metadata to the results.yml
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-test-it" class="md-nav__link">
    6. Test it.
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/monarch-initiative/pheval/edit/docs/docs/developing_a_pheval_plugin.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


<h1 id="developing-a-pheval-plugin">Developing a PhEval Plugin</h1>
<h2 id="description">Description</h2>
<p>Plugin development allows PhEval to be extensible, as we have designed it.
The plugin goal is to be flexible through custom runner implementations. This plugin development enhances the PhEval functionality. You can build one quickly using this step-by-step process.</p>
<p><mark><em>All custom Runners implementations must implement all</em> <strong>PhevalRunner</strong> <em>methods</em></mark></p>


<div class="doc doc-object doc-class">


<a id="src.pheval.runners.runner.PhEvalRunner"></a>
  <div class="doc doc-contents first">
      <p class="doc doc-class-bases">
        Bases: <code><span title="abc.ABC">ABC</span></code></p>

  
      <p>PhEvalRunner Class</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/runners/runner.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">PhEvalRunner</span><span class="p">(</span><span class="n">ABC</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;PhEvalRunner Class&quot;&quot;&quot;</span>

    <span class="n">input_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">testdata_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">tmp_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">config_file</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">directory_path</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">input_dir_config</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">_meta_data</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">__raw_results_dir</span> <span class="o">=</span> <span class="s2">&quot;raw_results/&quot;</span>
    <span class="n">__pheval_gene_results_dir</span> <span class="o">=</span> <span class="s2">&quot;pheval_gene_results/&quot;</span>
    <span class="n">__pheval_variant_results_dir</span> <span class="o">=</span> <span class="s2">&quot;pheval_variant_results/&quot;</span>
    <span class="n">__pheval_disease_results_dir</span> <span class="o">=</span> <span class="s2">&quot;pheval_disease_results/&quot;</span>
    <span class="n">__tool_input_commands_dir</span> <span class="o">=</span> <span class="s2">&quot;tool_input_commands/&quot;</span>
    <span class="n">__run_meta_data_file</span> <span class="o">=</span> <span class="s2">&quot;results.yml&quot;</span>

    <span class="k">def</span> <span class="nf">__post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span> <span class="o">=</span> <span class="n">parse_input_dir_config</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_dir</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_tool</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">tool</span>

    <span class="k">def</span> <span class="nf">_get_variant_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">variant_analysis</span>

    <span class="k">def</span> <span class="nf">_get_gene_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">gene_analysis</span>

    <span class="k">def</span> <span class="nf">_get_disease_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">disease_analysis</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">tool_input_commands_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__tool_input_commands_dir</span><span class="p">)</span>

    <span class="nd">@tool_input_commands_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">tool_input_commands_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">raw_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__raw_results_dir</span><span class="p">)</span>

    <span class="nd">@raw_results_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">raw_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">pheval_gene_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__pheval_gene_results_dir</span><span class="p">)</span>

    <span class="nd">@pheval_gene_results_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">pheval_gene_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">pheval_variant_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__pheval_variant_results_dir</span><span class="p">)</span>

    <span class="nd">@pheval_variant_results_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">pheval_variant_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">pheval_disease_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span><span class="p">)</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">__pheval_disease_results_dir</span><span class="p">)</span>

    <span class="nd">@pheval_disease_results_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">pheval_disease_results_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">directory_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory_path</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">build_output_directory_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;build output directory structure&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tool_input_commands_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">raw_results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_variant_analysis</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pheval_variant_results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_gene_analysis</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pheval_gene_results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_disease_analysis</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pheval_disease_results_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">meta_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_meta_data</span> <span class="o">=</span> <span class="n">BasicOutputRunMetaData</span><span class="p">(</span>
            <span class="n">tool</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">tool</span><span class="p">,</span>
            <span class="n">tool_version</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">version</span><span class="p">,</span>
            <span class="n">config</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_dir</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">run_timestamp</span><span class="o">=</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">timestamp</span><span class="p">(),</span>
            <span class="n">corpus</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">testdata_dir</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">testdata_dir</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_meta_data</span>

    <span class="nd">@meta_data</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">meta_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">meta_data</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_meta_data</span> <span class="o">=</span> <span class="n">meta_data</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">prepare</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;prepare&quot;&quot;&quot;</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;run&quot;&quot;&quot;</span>

    <span class="nd">@abstractmethod</span>
    <span class="k">def</span> <span class="nf">post_process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;post_process&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">construct_meta_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Construct run output meta data&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">meta_data</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div><hr />
<h2 id="step-by-step-plugin-development-process">Step-by-Step Plugin Development Process</h2>
<p>The plugin structure is derived from a <a href="https://cookiecutter.readthedocs.io/en/stable/">cookiecutter</a> template, <a href="https://github.com/hrshdhgd/sphintoxetry-cookiecutter">Sphintoxetry-cookiecutter</a>, and it uses <a href="https://www.sphinx-doc.org/en/master/">Sphinx</a>, <a href="https://tox.wiki/en/latest/">tox</a> and <a href="https://python-poetry.org">poetry</a> as core dependencies.
This allows PhEval extensibility to be standardized in terms of documentation and dependency management.</p>
<h3 id="1-sphintoxetry-cookiecutter-scaffold">1. Sphintoxetry-cookiecutter scaffold</h3>
<p>First, install the cruft package. Cruft enables keeping projects up-to-date with future updates made to this original template.</p>
<p>Install the latest release of cruft from pip</p>
<div class="highlight"><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>cruft
</code></pre></div>
<blockquote>
<p><strong><em>NOTE:</em></strong>  You may encounter an error with the naming of the project layout if using an older release of cruft. To avoid this, make sure you have installed the latest release version.</p>
</blockquote>
<p>Next, create a project using the sphintoxetry-cookiecutter template.</p>
<div class="highlight"><pre><span></span><code>cruft create https://github.com/monarch-initiative/monarch-project-template
</code></pre></div>
<h3 id="2-further-setup">2. Further setup</h3>
<h4 id="install-poetry-if-you-havent-already">Install poetry if you haven't already.</h4>
<div class="highlight"><pre><span></span><code>pip install poetry
</code></pre></div>
<h4 id="install-dependencies">Install dependencies</h4>
<div class="highlight"><pre><span></span><code>poetry install
</code></pre></div>
<h4 id="add-pheval-dependency">Add PhEval dependency</h4>
<div class="highlight"><pre><span></span><code>poetry add pheval
</code></pre></div>
<h4 id="run-tox-to-see-if-the-setup-works">Run tox to see if the setup works</h4>
<div class="highlight"><pre><span></span><code>poetry run tox
</code></pre></div>
<h3 id="3-implement-pheval-custom-runner">3. Implement PhEval Custom Runner</h3>
<p><em>The runner name is arbitrary and custom Runner name was chose by demonstrative purposes</em></p>
<p>Create a runner file inside the plugin project, e.g:</p>
<p><img alt="a" src="../imgs/plugin_example_folder_structure.png" /></p>
<div class="highlight"><pre><span></span><code><span class="sd">&quot;&quot;&quot;Custom Pheval Runner.&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">pheval.runners.runner</span> <span class="kn">import</span> <span class="n">PhEvalRunner</span>


<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">CustomPhevalRunner</span><span class="p">(</span><span class="n">PhEvalRunner</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;CustomPhevalRunner Class.&quot;&quot;&quot;</span>

    <span class="n">input_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">testdata_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">tmp_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">config_file</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span>

    <span class="k">def</span> <span class="nf">prepare</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;prepare method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;preparing&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;run method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;running with custom pheval runner&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">post_process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;post_process method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;post processing&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="4-add-pheval-plugins-section-to-the-pyprojecttoml-file">4. Add PhEval Plugins section to the pyproject.toml file</h3>
<div class="highlight"><pre><span></span><code><span class="k">[tool.poetry.plugins.&quot;pheval.plugins&quot;]</span>
<span class="n">customrunner</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;pheval_plugin_example.runner:CustomPhevalRunner&quot;</span>
</code></pre></div>
<p><mark>Replace the value above with the path to your custom runner plugin</mark></p>
<h3 id="5-implementing-pheval-helper-methods">5. Implementing PhEval helper methods</h3>
<p>Streamlining the creation of your custom PhEval runner can be facilitated by leveraging PhEval's versatile helper methods, where applicable.</p>
<p>Within PhEval, numerous public methods have been designed to assist in your runner methods. The utilisation of these helper methods is optional, yet they are crafted to enhance the overall implementation process.</p>
<h4 id="utility-methods">Utility methods</h4>
<p>The <code>PhenopacketUtil</code> class is designed to aid in the collection of specific data from a Phenopacket.</p>


<div class="doc doc-object doc-class">


<a id="src.pheval.utils.phenopacket_utils.PhenopacketUtil"></a>
  <div class="doc doc-contents first">

  
      <p>Class for retrieving data from a Phenopacket or Family object</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">PhenopacketUtil</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for retrieving data from a Phenopacket or Family object&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_contents</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialise PhenopacketUtil</span>

<span class="sd">        Args:</span>
<span class="sd">            phenopacket_contents (Union[Phenopacket, Family]): Phenopacket or Family object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span> <span class="o">=</span> <span class="n">phenopacket_contents</span>

    <span class="k">def</span> <span class="nf">sample_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the sample ID from a Phenopacket or proband of a Family</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Sample ID</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span> <span class="nf">phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">observed_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all observed HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of observed HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">negated_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all negated HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of negated HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">negated_phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="n">negated_phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">negated_phenotypic_features</span>

    <span class="k">def</span> <span class="nf">diseases</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Disease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of Diseases associated with the proband</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Disease]: List of diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">diseases</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">diseases</span>

    <span class="k">def</span> <span class="nf">_diagnosis_from_interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of disease diagnoses associated with the proband from the interpretations object</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diagnoses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
            <span class="p">(</span>
                <span class="n">diagnoses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">ProbandDisease</span><span class="p">(</span>
                        <span class="n">disease_name</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                        <span class="n">disease_identifier</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">label</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">id</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">diagnoses</span>

    <span class="k">def</span> <span class="nf">_diagnosis_from_disease</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of disease diagnoses associated with the proband from the diseases object</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diagnoses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">disease</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">diseases</span><span class="p">():</span>
            <span class="n">diagnoses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">ProbandDisease</span><span class="p">(</span><span class="n">disease_name</span><span class="o">=</span><span class="n">disease</span><span class="o">.</span><span class="n">term</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">disease_identifier</span><span class="o">=</span><span class="n">disease</span><span class="o">.</span><span class="n">term</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">diagnoses</span>

    <span class="k">def</span> <span class="nf">diagnoses</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a unique list of disease diagnoses associated with the proband from a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_interpretations</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_disease</span><span class="p">()))</span>

    <span class="k">def</span> <span class="nf">interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of interpretations from a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Interpretation]: List of interpretations</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">interpretations</span>

    <span class="k">def</span> <span class="nf">causative_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of causative variants listed in a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandCausativeVariant]: List of proband causative variants</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span>
                <span class="n">genotype</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">allelic_state</span>
                <span class="n">variant_data</span> <span class="o">=</span> <span class="n">ProbandCausativeVariant</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">genome_assembly</span><span class="p">,</span>
                    <span class="n">GenomicVariant</span><span class="p">(</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">genotype</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">all_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">all_variants</span>

    <span class="k">def</span> <span class="nf">files</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">File</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of files associated with a phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[File]: List of files associated with a phenopacket</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">files</span>

    <span class="k">def</span> <span class="nf">vcf_file_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">File</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the genome assembly and VCF file name from a phenopacket.</span>

<span class="sd">        Args:</span>
<span class="sd">            phenopacket_path (Path): The path to the phenopacket file.</span>
<span class="sd">            vcf_dir (Path): The directory path where the VCF file is stored.</span>

<span class="sd">        Returns:</span>
<span class="sd">            File: The VCF file with updated URI pointing to the specified directory.</span>

<span class="sd">        Raises:</span>
<span class="sd">            IncorrectFileFormatError: If the provided file is not in .vcf or .vcf.gz format.</span>
<span class="sd">            IncompatibleGenomeAssemblyError: If the genome assembly of the VCF file is not compatible.</span>

<span class="sd">        Note:</span>
<span class="sd">            This function searches for a VCF file within the provided list of files, validates its format,</span>
<span class="sd">            and checks if the genome assembly is compatible. If the conditions are met, it updates the</span>
<span class="sd">            URI of the VCF file to the specified directory and returns the modified file object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">compatible_genome_assembly</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">]</span>
        <span class="n">vcf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">()</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;vcf&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span>
            <span class="s2">&quot;.vcf.gz&quot;</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="n">IncorrectFileFormatError</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">),</span> <span class="s2">&quot;.vcf or .vcf.gz file&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">compatible_genome_assembly</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">IncompatibleGenomeAssemblyError</span><span class="p">(</span>
                <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">],</span> <span class="n">phenopacket_path</span>
            <span class="p">)</span>
        <span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">vcf_data</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_extract_diagnosed_gene</span><span class="p">(</span>
        <span class="n">genomic_interpretation</span><span class="p">:</span> <span class="n">GenomicInterpretation</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProbandCausativeGene</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the disease causing genes from the variant descriptor field if not empty,</span>
<span class="sd">        otherwise, retrieves from the gene descriptor from a phenopacket.</span>
<span class="sd">        Args:</span>
<span class="sd">            genomic_interpretation (GenomicInterpretation): A genomic interpretation from a Phenopacket</span>
<span class="sd">        Returns:</span>
<span class="sd">            ProbandCausativeGene: The disease causing gene</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">ByteSize</span><span class="p">()</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">gene_symbol</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">gene_identifier</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">diagnosed_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeGene</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the disease causing genes from a phenopacket.</span>
<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandCausativeGene]: List of causative genes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_diagnosed_gene</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
                <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">}</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">genes</span>

    <span class="k">def</span> <span class="nf">diagnosed_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">GenomicVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all known causative variants from a phenopacket.</span>
<span class="sd">        Returns:</span>
<span class="sd">            List[GenomicVariant]: List of causative variants</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">variant</span> <span class="o">=</span> <span class="n">GenomicVariant</span><span class="p">(</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">pos</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">ref</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                    <span class="n">alt</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div><hr />
<p><code>PhenopacketUtil</code> proves particularly beneficial in scenarios where the tool for which you're crafting a runner implementation does not directly accept Phenopackets as inputs. Instead, it might require elementssuch as HPO IDs via the command-line interface (CLI). In this context, leveraging PhenopacketUtil within the runner's preparation phase enables the extraction of observed phenotypic features from the Phenopacket input, facilitating seamless processing.</p>
<p>An example of how this could be implemented is outlined here:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pheval.utils.phenopacket_utils</span> <span class="kn">import</span> <span class="n">phenopacket_reader</span>
<span class="kn">from</span> <span class="nn">pheval.utils.phenopacket_utils</span> <span class="kn">import</span> <span class="n">PhenopacketUtil</span>

<span class="n">phenopacket</span> <span class="o">=</span> <span class="n">phenopacket_reader</span><span class="p">(</span><span class="s2">&quot;/path/to/phenopacket.json&quot;</span><span class="p">)</span>
<span class="n">phenopacket_util</span> <span class="o">=</span> <span class="n">PhenopacketUtil</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span>
<span class="c1"># To return a list of all observed phenotypes for a phenopacket</span>
<span class="n">observed_phenotypes</span> <span class="o">=</span> <span class="n">phenopacket_util</span><span class="o">.</span><span class="n">observed_phenotypic_features</span><span class="p">()</span>
<span class="c1"># To extract just the HPO ID as a list</span>
<span class="n">observed_phenotypes_hpo_ids</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">observed_phenotype</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">observed_phenotype</span> <span class="ow">in</span> <span class="n">observed_phenotypes</span>
<span class="p">]</span>
</code></pre></div>
<h4 id="additional-tool-specific-configurations">Additional tool-specific configurations</h4>
<p>For the <code>pheval run</code> command to execute successfully, a <code>config.yaml</code> should be found within the input directory supplied on the CLI.</p>
<div class="highlight"><pre><span></span><code><span class="nt">tool</span><span class="p">:</span><span class="w"> </span>
<span class="nt">tool_version</span><span class="p">:</span><span class="w"> </span>
<span class="nt">variant_analysis</span><span class="p">:</span><span class="w"> </span>
<span class="nt">gene_analysis</span><span class="p">:</span><span class="w"> </span>
<span class="nt">disease_analysis</span><span class="p">:</span><span class="w"> </span>
<span class="nt">tool_specific_configuration_options</span><span class="p">:</span>
</code></pre></div>
<p>The <code>tool_specific_configuration_options</code> is an optional field that can be populated with any variables specific to your runner implementation that is required for the running of your tool.</p>
<p>All other fields are required to be filled in. The <code>variant_analysis</code>, <code>gene_analysis</code>, and <code>disease_analysis</code> are set as booleans and are for specifying what type of analysis/prioritisation the tool outputs.</p>
<p>To populate the <code>tool_specific_configurations_options</code> with customised data, we suggest using the <code>pydantic</code> package as it can easily parse the data from the yaml structure.</p>
<p>e.g.,</p>
<p><em>Define a <code>BaseModel</code> class with the fields that will populate the <code>tool_specific_configuration_options</code></em></p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pydantic</span> <span class="kn">import</span> <span class="n">BaseModel</span><span class="p">,</span> <span class="n">Field</span>

<span class="k">class</span> <span class="nc">CustomisedConfigurations</span><span class="p">(</span><span class="n">BaseModel</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class for defining the customised configurations in tool_specific_configurations field,</span>
<span class="sd">    within the input_dir config.yaml</span>
<span class="sd">    Args:</span>
<span class="sd">        environment (str): Environment to run</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">environment</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">Field</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
</code></pre></div>
<p><em>Within your runner parse the field into an object.</em></p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">from</span> <span class="nn">pheval.runners.runner</span> <span class="kn">import</span> <span class="n">PhEvalRunner</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">CustomPhevalRunner</span><span class="p">(</span><span class="n">PhEvalRunner</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;CustomPhevalRunner Class.&quot;&quot;&quot;</span>

    <span class="n">input_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">testdata_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">tmp_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">config_file</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span>

    <span class="k">def</span> <span class="nf">prepare</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;prepare method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;preparing&quot;</span><span class="p">)</span>
        <span class="n">config</span> <span class="o">=</span> <span class="n">CustomisedConfigurations</span><span class="o">.</span><span class="n">parse_obj</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">input_dir_config</span><span class="o">.</span><span class="n">tool_specific_configuration_options</span>
        <span class="p">)</span>
        <span class="n">environment</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">environment</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;run method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;running with custom pheval runner&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">post_process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;post_process method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;post processing&quot;</span><span class="p">)</span>
</code></pre></div>
<h4 id="post-processing-methods">Post-processing methods</h4>
<p>PhEval currently supports the benchmarking of gene, variant, and disease prioritisation results. </p>
<p>To benchmark these result types, PhEval TSV result files need to be generated. </p>
<p>PhEval can deal with the ranking and generation of these files to the correct location. However, the runner implementation must handle the extraction of essential data from the tool-specific raw results. This involves transforming them into a list comprising PhEval data classes, with each instance representing a result entry.</p>
<p>The dataclasses representing essential information extracted from tool-specific output for gene, variant, and disease prioritisation are defined as follows:</p>


<div class="doc doc-object doc-class">


<a id="src.pheval.post_processing.post_processing.PhEvalGeneResult"></a>
  <div class="doc doc-contents first">
      <p class="doc doc-class-bases">
        Bases: <code><a class="autorefs autorefs-internal" title="src.pheval.post_processing.post_processing.PhEvalResult" href="../api/pheval/post_processing/post_processing/#src.pheval.post_processing.post_processing.PhEvalResult">PhEvalResult</a></code></p>

  
      <p>Minimal data required from tool-specific output for gene prioritisation result</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene_symbol</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The gene symbol for the result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gene_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The ENSEMBL gene identifier for the result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>score</code></td>
          <td>
                <code>float</code>
          </td>
          <td><p>The score for the gene result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>While we recommend providing the gene identifier in the ENSEMBL namespace,
any matching format used in Phenopacket interpretations is acceptable for result matching purposes
in the analysis.</p>
</details>

        <details class="quote">
          <summary>Source code in <code>src/pheval/post_processing/post_processing.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">PhEvalGeneResult</span><span class="p">(</span><span class="n">PhEvalResult</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Minimal data required from tool-specific output for gene prioritisation result</span>
<span class="sd">    Args:</span>
<span class="sd">        gene_symbol (str): The gene symbol for the result entry</span>
<span class="sd">        gene_identifier (str): The ENSEMBL gene identifier for the result entry</span>
<span class="sd">        score (float): The score for the gene result entry</span>
<span class="sd">    Notes:</span>
<span class="sd">        While we recommend providing the gene identifier in the ENSEMBL namespace,</span>
<span class="sd">        any matching format used in Phenopacket interpretations is acceptable for result matching purposes</span>
<span class="sd">        in the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_identifier</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">score</span><span class="p">:</span> <span class="nb">float</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div><hr />


<div class="doc doc-object doc-class">


<a id="src.pheval.post_processing.post_processing.PhEvalVariantResult"></a>
  <div class="doc doc-contents first">
      <p class="doc doc-class-bases">
        Bases: <code><a class="autorefs autorefs-internal" title="src.pheval.post_processing.post_processing.PhEvalResult" href="../api/pheval/post_processing/post_processing/#src.pheval.post_processing.post_processing.PhEvalResult">PhEvalResult</a></code></p>

  
      <p>Minimal data required from tool-specific output for variant prioritisation</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>chromosome</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The chromosome position of the variant recommended to be provided in the following format.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>start</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>The start position of the variant</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>end</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>The end position of the variant</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>ref</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The reference allele of the variant</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>alt</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The alternate allele of the variant</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>score</code></td>
          <td>
                <code>float</code>
          </td>
          <td><p>The score for the variant result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>While we recommend providing the variant's chromosome in the specified format,
any matching format used in Phenopacket interpretations is acceptable for result matching purposes
in the analysis.</p>
</details>

        <details class="quote">
          <summary>Source code in <code>src/pheval/post_processing/post_processing.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">PhEvalVariantResult</span><span class="p">(</span><span class="n">PhEvalResult</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Minimal data required from tool-specific output for variant prioritisation</span>
<span class="sd">    Args:</span>
<span class="sd">        chromosome (str): The chromosome position of the variant recommended to be provided in the following format.</span>
<span class="sd">        This includes numerical designations from 1 to 22 representing autosomal chromosomes,</span>
<span class="sd">        as well as the sex chromosomes X and Y, and the mitochondrial chromosome MT.</span>
<span class="sd">        start (int): The start position of the variant</span>
<span class="sd">        end (int): The end position of the variant</span>
<span class="sd">        ref (str): The reference allele of the variant</span>
<span class="sd">        alt (str): The alternate allele of the variant</span>
<span class="sd">        score (float): The score for the variant result entry</span>
<span class="sd">    Notes:</span>
<span class="sd">        While we recommend providing the variant&#39;s chromosome in the specified format,</span>
<span class="sd">        any matching format used in Phenopacket interpretations is acceptable for result matching purposes</span>
<span class="sd">        in the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">ref</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">alt</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">score</span><span class="p">:</span> <span class="nb">float</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div><hr />


<div class="doc doc-object doc-class">


<a id="src.pheval.post_processing.post_processing.PhEvalDiseaseResult"></a>
  <div class="doc doc-contents first">
      <p class="doc doc-class-bases">
        Bases: <code><a class="autorefs autorefs-internal" title="src.pheval.post_processing.post_processing.PhEvalResult" href="../api/pheval/post_processing/post_processing/#src.pheval.post_processing.post_processing.PhEvalResult">PhEvalResult</a></code></p>

  
      <p>Minimal data required from tool-specific output for disease prioritisation</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>disease_name</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Disease name for the result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>disease_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Identifier for the disease result entry in the OMIM namespace</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>score</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Score for the disease result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>While we recommend providing the disease identifier in the OMIM namespace,
any matching format used in Phenopacket interpretations is acceptable for result matching purposes
in the analysis.</p>
</details>

        <details class="quote">
          <summary>Source code in <code>src/pheval/post_processing/post_processing.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">PhEvalDiseaseResult</span><span class="p">(</span><span class="n">PhEvalResult</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Minimal data required from tool-specific output for disease prioritisation</span>
<span class="sd">    Args:</span>
<span class="sd">        disease_name (str): Disease name for the result entry</span>
<span class="sd">        disease_identifier (str): Identifier for the disease result entry in the OMIM namespace</span>
<span class="sd">        score (str): Score for the disease result entry</span>
<span class="sd">    Notes:</span>
<span class="sd">        While we recommend providing the disease identifier in the OMIM namespace,</span>
<span class="sd">        any matching format used in Phenopacket interpretations is acceptable for result matching purposes</span>
<span class="sd">        in the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">disease_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">disease_identifier</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">score</span><span class="p">:</span> <span class="nb">float</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div><hr />
<p>The <code>generate_pheval_result()</code> can be implemented in your runner to write out the PhEval TSV results.</p>
<p>An example of how the method can be called is outlined here:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pheval.post_processing.post_processing</span> <span class="kn">import</span> <span class="n">generate_pheval_result</span>

<span class="n">generate_pheval_result</span><span class="p">(</span>
    <span class="n">pheval_result</span><span class="o">=</span><span class="n">pheval_gene_result</span><span class="p">,</span> <span class="c1"># this is the list of extracted PhEval result requirements</span>
    <span class="n">sort_order_str</span><span class="o">=</span><span class="s2">&quot;descending&quot;</span><span class="p">,</span> <span class="c1"># or can be ascending - this determines in which order the scores will be ranked</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="n">output_directory</span><span class="p">,</span> <span class="c1"># this can be accessed from the runner instance e.g., self.output_dir</span>
    <span class="n">tool_result_path</span><span class="o">=</span><span class="n">tool_result_json</span> <span class="c1"># this is the path to the tool-specific raw results file</span>
<span class="p">)</span>
</code></pre></div>
<h4 id="adding-metadata-to-the-resultsyml">Adding metadata to the results.yml</h4>
<p>By default, PhEval will write a <code>results.yml</code> to the output directory supplied on the CLI. </p>
<p>The <code>results.yml</code> contains basic metadata regarding the run configuration, however, there is also the option to add customised run metadata to the <code>results.yml</code> in the <code>tool_specific_configuration_options</code> field.</p>
<p>To achieve this, you'll need to create a <code>construct_meta_data()</code> method within your runner implementation. This method is responsible for appending customised metadata to the metadata object in the form of a defined dataclass. It should return the entire metadata object once the addition is completed.</p>
<p>e.g.,</p>
<p><em>Defined customised metadata dataclass:</em></p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>

<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">CustomisedMetaData</span><span class="p">:</span>
    <span class="n">customised_field</span><span class="p">:</span> <span class="nb">str</span>
</code></pre></div>
<p><em>Example of implementation in the runner.</em></p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">from</span> <span class="nn">pheval.runners.runner</span> <span class="kn">import</span> <span class="n">PhEvalRunner</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">CustomPhevalRunner</span><span class="p">(</span><span class="n">PhEvalRunner</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;CustomPhevalRunner Class.&quot;&quot;&quot;</span>

    <span class="n">input_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">testdata_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">tmp_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">config_file</span><span class="p">:</span> <span class="n">Path</span>
    <span class="n">version</span><span class="p">:</span> <span class="nb">str</span>

    <span class="k">def</span> <span class="nf">prepare</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;prepare method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;preparing&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;run method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;running with custom pheval runner&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">post_process</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;post_process method.&quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;post processing&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">construct_meta_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add metadata.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">meta_data</span><span class="o">.</span><span class="n">tool_specific_configuration_options</span> <span class="o">=</span> <span class="n">CustomisedMetaData</span><span class="p">(</span><span class="n">customised_field</span><span class="o">=</span><span class="s2">&quot;customised_value&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">meta_data</span>
</code></pre></div>
<h3 id="6-test-it">6. Test it.</h3>
<p>To update your custom pheval runner implementation, you must first install the package</p>
<div class="highlight"><pre><span></span><code>poetry install
</code></pre></div>
<p>Now you have to be able to run PhEval passing your custom runner as parameter. e.g.,</p>
<div class="highlight"><pre><span></span><code>pheval run -i ./input_dir -t ./test_data_dir -r &#39;customphevalrunner&#39; -o output_dir
</code></pre></div>
<p>The <code>-r</code> parameter stands for your plugin runner class name, and it must be entirely lowercase.</p>
<p>Output:</p>
<div class="highlight"><pre><span></span><code>preparing
running with custom pheval Runner
post processing
</code></pre></div>
<p>Pay attention to "<em><mark>running with custom pheval Runner</mark></em>" line, this is exactly what we had implemented in the <strong>CustomPhevalRunner</strong> Example</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../api/pheval/utils/utils/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Utils" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Utils
            </div>
          </div>
        </a>
      
      
        
        <a href="../contributing/" class="md-footer__link md-footer__link--next" aria-label="Next: Contributions" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Contributions
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.5a2dcb6a.min.js"></script>
      
        <script src="https://unpkg.com/mermaid@8.5.1/dist/mermaid.min.js"></script>
      
    
    
  </body>
</html>