
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/pheval/api/pheval/utils/phenopacket_utils/">
      
      <link rel="icon" href="../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-8.5.10">
    
    
      
        <title>Phenopacket utils - PhEval</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.975780f9.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.2505c338.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/mermaid@8.5.1/dist/mermaid.css">
    
      <link rel="stylesheet" href="../../../../mermaid.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.pheval.utils.phenopacket_utils" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="PhEval" class="md-header__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PhEval
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Phenopacket utils
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Night Mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Night Mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
            </label>
          
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Day Mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Day Mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="PhEval" class="md-nav__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    PhEval
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" >
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Pipeline
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Pipeline" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pipeline
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../exomiser_pipeline/" class="md-nav__link">
        PhEval Pipeline Exomiser Runner
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../pipeline/" class="md-nav__link">
        PhEval Pipeline
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../contact/" class="md-nav__link">
        Contact Us
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          API Documentation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API Documentation" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          API Documentation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/" class="md-nav__link">
        Cli
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../config_parser/" class="md-nav__link">
        Config parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../constants/" class="md-nav__link">
        Constants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../run_metadata/" class="md-nav__link">
        Run metadata
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_5" type="checkbox" id="__nav_5_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_5">
          analyse
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="analyse" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_5">
          <span class="md-nav__icon md-icon"></span>
          analyse
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/analysis/" class="md-nav__link">
        Analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/benchmark_generator/" class="md-nav__link">
        Benchmark generator
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/benchmarking_data/" class="md-nav__link">
        Benchmarking data
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/disease_prioritisation_analysis/" class="md-nav__link">
        Disease prioritisation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/gene_prioritisation_analysis/" class="md-nav__link">
        Gene prioritisation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/generate_plots/" class="md-nav__link">
        Generate plots
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/generate_summary_outputs/" class="md-nav__link">
        Generate summary outputs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/parse_benchmark_summary/" class="md-nav__link">
        Parse benchmark summary
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/parse_pheval_result/" class="md-nav__link">
        Parse pheval result
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/prioritisation_rank_recorder/" class="md-nav__link">
        Prioritisation rank recorder
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/prioritisation_result_types/" class="md-nav__link">
        Prioritisation result types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/rank_stats/" class="md-nav__link">
        Rank stats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/run_data_parser/" class="md-nav__link">
        Run data parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../analyse/variant_prioritisation_analysis/" class="md-nav__link">
        Variant prioritisation analysis
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_6" type="checkbox" id="__nav_5_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_6">
          post_processing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="post_processing" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_6">
          <span class="md-nav__icon md-icon"></span>
          post_processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../post_processing/post_processing/" class="md-nav__link">
        Post processing
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_7" type="checkbox" id="__nav_5_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_7">
          prepare
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="prepare" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_7">
          <span class="md-nav__icon md-icon"></span>
          prepare
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/create_noisy_phenopackets/" class="md-nav__link">
        Create noisy phenopackets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/create_spiked_vcf/" class="md-nav__link">
        Create spiked vcf
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/custom_exceptions/" class="md-nav__link">
        Custom exceptions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../prepare/update_phenopacket/" class="md-nav__link">
        Update phenopacket
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_8" type="checkbox" id="__nav_5_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5_8">
          runners
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="runners" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_8">
          <span class="md-nav__icon md-icon"></span>
          runners
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../runners/runner/" class="md-nav__link">
        Runner
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5_9" type="checkbox" id="__nav_5_9" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5_9">
          utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="utils" data-md-level="2">
        <label class="md-nav__title" for="__nav_5_9">
          <span class="md-nav__icon md-icon"></span>
          utils
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../file_utils/" class="md-nav__link">
        File utils
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Phenopacket utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Phenopacket utils
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils" class="md-nav__link">
    src.pheval.utils.phenopacket_utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="md-nav__link">
    GeneIdentifierUpdater
  </a>
  
    <nav class="md-nav" aria-label="GeneIdentifierUpdater">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="md-nav__link">
    find_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="md-nav__link">
    obtain_gene_symbol_from_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="md-nav__link">
    update_genomic_interpretations_gene_identifier()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GenomicVariant" class="md-nav__link">
    GenomicVariant
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="md-nav__link">
    IncompatibleGenomeAssemblyError
  </a>
  
    <nav class="md-nav" aria-label="IncompatibleGenomeAssemblyError">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="md-nav__link">
    PhenopacketRebuilder
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketRebuilder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="md-nav__link">
    add_randomised_hpo()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="md-nav__link">
    add_spiked_vcf_path()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="md-nav__link">
    update_interpretations()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="md-nav__link">
    causative_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="md-nav__link">
    diagnosed_genes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="md-nav__link">
    diagnosed_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnoses" class="md-nav__link">
    diagnoses()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diseases" class="md-nav__link">
    diseases()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="md-nav__link">
    files()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="md-nav__link">
    interpretations()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="md-nav__link">
    negated_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="md-nav__link">
    observed_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="md-nav__link">
    phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="md-nav__link">
    sample_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="md-nav__link">
    vcf_file_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandCausativeGene" class="md-nav__link">
    ProbandCausativeGene
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandCausativeVariant" class="md-nav__link">
    ProbandCausativeVariant
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandDisease" class="md-nav__link">
    ProbandDisease
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_gene_identifier_map" class="md-nav__link">
    create_gene_identifier_map()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="md-nav__link">
    create_hgnc_dict()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_json_message" class="md-nav__link">
    create_json_message()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.phenopacket_reader" class="md-nav__link">
    phenopacket_reader()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.read_hgnc_data" class="md-nav__link">
    read_hgnc_data()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.write_phenopacket" class="md-nav__link">
    write_phenopacket()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../semsim_utils/" class="md-nav__link">
        Semsim utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        Utils
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Resources for contributors
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Resources for contributors" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Resources for contributors
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../developing_a_pheval_plugin/" class="md-nav__link">
        Developing a PhEval Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../contributing/" class="md-nav__link">
        Contributions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../styleguide/" class="md-nav__link">
        Monarch Style Guide for PhEval
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../../CODE_OF_CONDUCT/" class="md-nav__link">
        Contributor Covenant Code of Conduct
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../../roadmap/" class="md-nav__link">
        Roadmap
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils" class="md-nav__link">
    src.pheval.utils.phenopacket_utils
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="md-nav__link">
    GeneIdentifierUpdater
  </a>
  
    <nav class="md-nav" aria-label="GeneIdentifierUpdater">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="md-nav__link">
    find_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="md-nav__link">
    obtain_gene_symbol_from_identifier()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="md-nav__link">
    update_genomic_interpretations_gene_identifier()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.GenomicVariant" class="md-nav__link">
    GenomicVariant
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="md-nav__link">
    IncompatibleGenomeAssemblyError
  </a>
  
    <nav class="md-nav" aria-label="IncompatibleGenomeAssemblyError">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="md-nav__link">
    PhenopacketRebuilder
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketRebuilder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="md-nav__link">
    add_randomised_hpo()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="md-nav__link">
    add_spiked_vcf_path()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="md-nav__link">
    update_interpretations()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="md-nav__link">
    PhenopacketUtil
  </a>
  
    <nav class="md-nav" aria-label="PhenopacketUtil">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.__init__" class="md-nav__link">
    __init__()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="md-nav__link">
    causative_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="md-nav__link">
    diagnosed_genes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="md-nav__link">
    diagnosed_variants()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnoses" class="md-nav__link">
    diagnoses()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.diseases" class="md-nav__link">
    diseases()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="md-nav__link">
    files()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="md-nav__link">
    interpretations()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="md-nav__link">
    negated_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="md-nav__link">
    observed_phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="md-nav__link">
    phenotypic_features()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="md-nav__link">
    sample_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="md-nav__link">
    vcf_file_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandCausativeGene" class="md-nav__link">
    ProbandCausativeGene
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandCausativeVariant" class="md-nav__link">
    ProbandCausativeVariant
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.ProbandDisease" class="md-nav__link">
    ProbandDisease
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_gene_identifier_map" class="md-nav__link">
    create_gene_identifier_map()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="md-nav__link">
    create_hgnc_dict()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.create_json_message" class="md-nav__link">
    create_json_message()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.phenopacket_reader" class="md-nav__link">
    phenopacket_reader()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.read_hgnc_data" class="md-nav__link">
    read_hgnc_data()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.pheval.utils.phenopacket_utils.write_phenopacket" class="md-nav__link">
    write_phenopacket()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/monarch-initiative/pheval/edit/docs/docs/api/pheval/utils/phenopacket_utils.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>Phenopacket utils</h1>

<div class="doc doc-object doc-module">


<a id="src.pheval.utils.phenopacket_utils"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater" class="doc doc-heading">
        <code>GeneIdentifierUpdater</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Class for updating gene identifiers within genomic interpretations.</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GeneIdentifierUpdater</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for updating gene identifiers within genomic interpretations.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hgnc_data</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">identifier_map</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialise the GeneIdentifierUpdater.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_identifier (str): The gene identifier to update to.</span>
<span class="sd">            hgnc_data (dict): A dictionary containing HGNC data (default: None).</span>
<span class="sd">            identifier_map (dict): A dictionary mapping gene identifiers (default: None).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span> <span class="o">=</span> <span class="n">hgnc_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span> <span class="o">=</span> <span class="n">gene_identifier</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span> <span class="o">=</span> <span class="n">identifier_map</span>

    <span class="k">def</span> <span class="nf">find_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find the specified gene identifier for a gene symbol.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_symbol (str): The gene symbol to find the identifier for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The identified gene identifier.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">_symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                        <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">query_gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtain gene symbol from a gene identifier.</span>

<span class="sd">        Args:</span>
<span class="sd">            query_gene_identifier (str): The gene identifier.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The gene symbol corresponding to the identifier.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span><span class="p">[</span><span class="n">query_gene_identifier</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">_find_alternate_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find the alternate IDs for a gene symbol.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_symbol (str): The gene symbol to find alternate IDs for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: List of alternate IDs for the gene symbol.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;ncbigene:&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;ensembl:&quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">],</span>
                <span class="s2">&quot;symbol:&quot;</span> <span class="o">+</span> <span class="n">gene_symbol</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                        <span class="k">return</span> <span class="p">[</span>
                            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;ncbigene:&quot;</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;ensembl:&quot;</span> <span class="o">+</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;symbol:&quot;</span> <span class="o">+</span> <span class="n">symbol</span><span class="p">,</span>
                        <span class="p">]</span>

    <span class="k">def</span> <span class="nf">update_genomic_interpretations_gene_identifier</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Update the genomic interpretations of a Phenopacket.</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretations (List[Interpretation]): List of Interpretation objects.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Interpretation]: Updated list of Interpretation objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">updated_interpretations</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">interpretations</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">updated_interpretation</span> <span class="ow">in</span> <span class="n">updated_interpretations</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">updated_interpretation</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">find_identifier</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">del</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="p">[:]</span>
                <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_find_alternate_ids</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                    <span class="p">)</span>
                <span class="p">)</span>
        <span class="k">return</span> <span class="n">updated_interpretations</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">gene_identifier</span><span class="p">,</span> <span class="n">hgnc_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">identifier_map</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Initialise the GeneIdentifierUpdater.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The gene identifier to update to.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>hgnc_data</code></td>
          <td>
                <code>dict</code>
          </td>
          <td><p>A dictionary containing HGNC data (default: None).</p></td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>identifier_map</code></td>
          <td>
                <code>dict</code>
          </td>
          <td><p>A dictionary mapping gene identifiers (default: None).</p></td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">hgnc_data</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">identifier_map</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialise the GeneIdentifierUpdater.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_identifier (str): The gene identifier to update to.</span>
<span class="sd">        hgnc_data (dict): A dictionary containing HGNC data (default: None).</span>
<span class="sd">        identifier_map (dict): A dictionary mapping gene identifiers (default: None).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span> <span class="o">=</span> <span class="n">hgnc_data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span> <span class="o">=</span> <span class="n">gene_identifier</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span> <span class="o">=</span> <span class="n">identifier_map</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.find_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">find_identifier</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Find the specified gene identifier for a gene symbol.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene_symbol</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The gene symbol to find the identifier for.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>str</code></td>          <td>
                <code>str</code>
          </td>
          <td><p>The identified gene identifier.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the specified gene identifier for a gene symbol.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_symbol (str): The gene symbol to find the identifier for.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The identified gene identifier.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="p">[</span><span class="n">gene_symbol</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_symbol</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hgnc_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">prev_symbol</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">prev_symbol</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_identifier</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.obtain_gene_symbol_from_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="n">query_gene_identifier</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Obtain gene symbol from a gene identifier.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>query_gene_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The gene identifier.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>str</code></td>          <td>
                <code>str</code>
          </td>
          <td><p>The gene symbol corresponding to the identifier.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">obtain_gene_symbol_from_identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">query_gene_identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Obtain gene symbol from a gene identifier.</span>

<span class="sd">    Args:</span>
<span class="sd">        query_gene_identifier (str): The gene identifier.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The gene symbol corresponding to the identifier.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">identifier_map</span><span class="p">[</span><span class="n">query_gene_identifier</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.GeneIdentifierUpdater.update_genomic_interpretations_gene_identifier" class="doc doc-heading">
<code class="highlight language-python"><span class="n">update_genomic_interpretations_gene_identifier</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Update the genomic interpretations of a Phenopacket.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>interpretations</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.Interpretation">Interpretation</span>]</code>
          </td>
          <td><p>List of Interpretation objects.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.Interpretation">Interpretation</span>]</code>
          </td>
          <td><p>List[Interpretation]: Updated list of Interpretation objects.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_genomic_interpretations_gene_identifier</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update the genomic interpretations of a Phenopacket.</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretations (List[Interpretation]): List of Interpretation objects.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[Interpretation]: Updated list of Interpretation objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">updated_interpretations</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">interpretations</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">updated_interpretation</span> <span class="ow">in</span> <span class="n">updated_interpretations</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">updated_interpretation</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">find_identifier</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">del</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="p">[:]</span>
            <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">alternate_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_find_alternate_ids</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">updated_interpretations</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.GenomicVariant" class="doc doc-heading">
        <code>GenomicVariant</code>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

</h2>


  <div class="doc doc-contents ">

  
      <p>Represents a genomic variant.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>chrom</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The chromosome position of the variant recommended to be provided in the following format.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>pos</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>Position of the variant following VCF convention.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>ref</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Reference allele following VCF convention.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>alt</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Alternate allele following VCF convention.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">GenomicVariant</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents a genomic variant.</span>

<span class="sd">    Args:</span>
<span class="sd">        chrom (str): The chromosome position of the variant recommended to be provided in the following format.</span>
<span class="sd">        This includes numerical designations from 1 to 22 representing autosomal chromosomes,</span>
<span class="sd">        as well as the sex chromosomes X and Y, and the mitochondrial chromosome MT.</span>
<span class="sd">        pos (int): Position of the variant following VCF convention.</span>
<span class="sd">        ref (str): Reference allele following VCF convention.</span>
<span class="sd">        alt (str): Alternate allele following VCF convention.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">pos</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">ref</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">alt</span><span class="p">:</span> <span class="nb">str</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" class="doc doc-heading">
        <code>IncompatibleGenomeAssemblyError</code>


</h2>


  <div class="doc doc-contents ">
      <p class="doc doc-class-bases">
        Bases: <code>Exception</code></p>

  
      <p>Exception raised for incompatible genome assembly.</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">IncompatibleGenomeAssemblyError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Exception raised for incompatible genome assembly.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Incompatible Genome Assembly&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialise IncompatibleGenomeAssemblyError.</span>

<span class="sd">        Attributes:</span>
<span class="sd">           assembly (str): Incompatible genome assembly encountered.</span>
<span class="sd">           phenopacket (Path): Path to the Phenopacket associated with the error.</span>
<span class="sd">           message (str, optional): Custom error message (default is &quot;Incompatible Genome Assembly&quot;).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">assembly</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">phenopacket</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">message</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="si">}</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">assembly</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s1">&#39;Incompatible Genome Assembly&#39;</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Initialise IncompatibleGenomeAssemblyError.</p>

  <p><strong>Attributes:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>assembly</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Incompatible genome assembly encountered.</p></td>
        </tr>
        <tr>
          <td><code>phenopacket</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td><p>Path to the Phenopacket associated with the error.</p></td>
        </tr>
        <tr>
          <td><code>message</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Custom error message (default is "Incompatible Genome Assembly").</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;Incompatible Genome Assembly&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialise IncompatibleGenomeAssemblyError.</span>

<span class="sd">    Attributes:</span>
<span class="sd">       assembly (str): Incompatible genome assembly encountered.</span>
<span class="sd">       phenopacket (Path): Path to the Phenopacket associated with the error.</span>
<span class="sd">       message (str, optional): Custom error message (default is &quot;Incompatible Genome Assembly&quot;).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">assembly</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">phenopacket</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">message</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder" class="doc doc-heading">
        <code>PhenopacketRebuilder</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Class for rebuilding a Phenopacket</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">PhenopacketRebuilder</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for rebuilding a Phenopacket&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialise PhenopacketUtil</span>

<span class="sd">        Attributes:</span>
<span class="sd">            phenopacket (Union[Phenopacket, Family]): Phenopacket or Family object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span> <span class="o">=</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">update_interpretations</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add the updated interpretations to a Phenopacket or Family.</span>

<span class="sd">        Args:</span>
<span class="sd">            interpretations (List[Interpretation]): The updated interpretations to be added.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Union[Phenopacket, Family]: The Phenopacket or Family object with updated interpretations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">add_randomised_hpo</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">randomised_hpo</span><span class="p">:</span> <span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add randomised phenotypic profiles to a Phenopacket or Family.</span>

<span class="sd">        Args:</span>
<span class="sd">            randomised_hpo: The randomised phenotypic profiles to be added.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Union[Phenopacket, Family] The Phenopacket or Family object with added randomised profiles.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
            <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>

    <span class="k">def</span> <span class="nf">add_spiked_vcf_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spiked_vcf_file_data</span><span class="p">:</span> <span class="n">File</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add a spiked VCF path to a Phenopacket or Family.</span>

<span class="sd">        Args:</span>
<span class="sd">        - spiked_vcf_file_data (File): The VCF file data to be added.</span>

<span class="sd">        Returns:</span>
<span class="sd">        - Phenopacket or Family: The Phenopacket or Family object with the added spiked VCF path.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
        <span class="n">phenopacket_files</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;vcf&quot;</span>
        <span class="p">]</span>
        <span class="n">phenopacket_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">phenopacket_files</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Initialise PhenopacketUtil</p>

  <p><strong>Attributes:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>phenopacket</code></td>
          <td>
                <code><span title="typing.Union">Union</span>[<span title="phenopackets.Phenopacket">Phenopacket</span>, <span title="phenopackets.Family">Family</span>]</code>
          </td>
          <td><p>Phenopacket or Family object</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialise PhenopacketUtil</span>

<span class="sd">    Attributes:</span>
<span class="sd">        phenopacket (Union[Phenopacket, Family]): Phenopacket or Family object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span> <span class="o">=</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_randomised_hpo" class="doc doc-heading">
<code class="highlight language-python"><span class="n">add_randomised_hpo</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Add randomised phenotypic profiles to a Phenopacket or Family.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>randomised_hpo</code></td>
          <td>
                <code>[<span title="phenopackets.PhenotypicFeature">PhenotypicFeature</span>]</code>
          </td>
          <td><p>The randomised phenotypic profiles to be added.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Union">Union</span>[<span title="phenopackets.Phenopacket">Phenopacket</span>, <span title="phenopackets.Family">Family</span>]</code>
          </td>
          <td><p>Union[Phenopacket, Family] The Phenopacket or Family object with added randomised profiles.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_randomised_hpo</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">randomised_hpo</span><span class="p">:</span> <span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add randomised phenotypic profiles to a Phenopacket or Family.</span>

<span class="sd">    Args:</span>
<span class="sd">        randomised_hpo: The randomised phenotypic profiles to be added.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Union[Phenopacket, Family] The Phenopacket or Family object with added randomised profiles.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">randomised_hpo</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.add_spiked_vcf_path" class="doc doc-heading">
<code class="highlight language-python"><span class="n">add_spiked_vcf_path</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Add a spiked VCF path to a Phenopacket or Family.</p>
      <ul>
<li>spiked_vcf_file_data (File): The VCF file data to be added.</li>
</ul>
      <ul>
<li>Phenopacket or Family: The Phenopacket or Family object with the added spiked VCF path.</li>
</ul>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_spiked_vcf_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spiked_vcf_file_data</span><span class="p">:</span> <span class="n">File</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add a spiked VCF path to a Phenopacket or Family.</span>

<span class="sd">    Args:</span>
<span class="sd">    - spiked_vcf_file_data (File): The VCF file data to be added.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Phenopacket or Family: The Phenopacket or Family object with the added spiked VCF path.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="n">phenopacket_files</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;vcf&quot;</span>
    <span class="p">]</span>
    <span class="n">phenopacket_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spiked_vcf_file_data</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="p">[:]</span>
    <span class="n">phenopacket</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">phenopacket_files</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketRebuilder.update_interpretations" class="doc doc-heading">
<code class="highlight language-python"><span class="n">update_interpretations</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Add the updated interpretations to a Phenopacket or Family.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>interpretations</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.Interpretation">Interpretation</span>]</code>
          </td>
          <td><p>The updated interpretations to be added.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Union">Union</span>[<span title="phenopackets.Phenopacket">Phenopacket</span>, <span title="phenopackets.Family">Family</span>]</code>
          </td>
          <td><p>Union[Phenopacket, Family]: The Phenopacket or Family object with updated interpretations.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">update_interpretations</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">interpretations</span><span class="p">:</span> <span class="p">[</span><span class="n">Interpretation</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add the updated interpretations to a Phenopacket or Family.</span>

<span class="sd">    Args:</span>
<span class="sd">        interpretations (List[Interpretation]): The updated interpretations to be added.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Union[Phenopacket, Family]: The Phenopacket or Family object with updated interpretations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="p">[:]</span>
        <span class="n">phenopacket</span><span class="o">.</span><span class="n">interpretations</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">interpretations</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenopacket</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil" class="doc doc-heading">
        <code>PhenopacketUtil</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Class for retrieving data from a Phenopacket or Family object</p>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">PhenopacketUtil</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for retrieving data from a Phenopacket or Family object&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_contents</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialise PhenopacketUtil</span>

<span class="sd">        Args:</span>
<span class="sd">            phenopacket_contents (Union[Phenopacket, Family]): Phenopacket or Family object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span> <span class="o">=</span> <span class="n">phenopacket_contents</span>

    <span class="k">def</span> <span class="nf">sample_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the sample ID from a Phenopacket or proband of a Family</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Sample ID</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span> <span class="nf">phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">observed_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all observed HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of observed HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">phenotypic_features</span>

    <span class="k">def</span> <span class="nf">negated_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all negated HPO terms</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[PhenotypicFeature]: List of negated HPO terms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">negated_phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
                <span class="n">negated_phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">negated_phenotypic_features</span>

    <span class="k">def</span> <span class="nf">diseases</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Disease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of Diseases associated with the proband</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Disease]: List of diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">diseases</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">diseases</span>

    <span class="k">def</span> <span class="nf">_diagnosis_from_interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of disease diagnoses associated with the proband from the interpretations object</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diagnoses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
            <span class="n">diagnoses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">ProbandDisease</span><span class="p">(</span>
                    <span class="n">disease_name</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                    <span class="n">disease_identifier</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">label</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">disease</span><span class="o">.</span><span class="n">id</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">diagnoses</span>

    <span class="k">def</span> <span class="nf">_diagnosis_from_disease</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of disease diagnoses associated with the proband from the diseases object</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">diagnoses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">disease</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">diseases</span><span class="p">():</span>
            <span class="n">diagnoses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">ProbandDisease</span><span class="p">(</span><span class="n">disease_name</span><span class="o">=</span><span class="n">disease</span><span class="o">.</span><span class="n">term</span><span class="o">.</span><span class="n">label</span><span class="p">,</span> <span class="n">disease_identifier</span><span class="o">=</span><span class="n">disease</span><span class="o">.</span><span class="n">term</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">diagnoses</span>

    <span class="k">def</span> <span class="nf">diagnoses</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a unique list of disease diagnoses associated with the proband from a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_interpretations</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_disease</span><span class="p">()))</span>

    <span class="k">def</span> <span class="nf">interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of interpretations from a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[Interpretation]: List of interpretations</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">interpretations</span>

    <span class="k">def</span> <span class="nf">causative_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of causative variants listed in a Phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandCausativeVariant]: List of proband causative variants</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span>
                <span class="n">genotype</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">allelic_state</span>
                <span class="n">variant_data</span> <span class="o">=</span> <span class="n">ProbandCausativeVariant</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">genome_assembly</span><span class="p">,</span>
                    <span class="n">GenomicVariant</span><span class="p">(</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                        <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">genotype</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">all_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">all_variants</span>

    <span class="k">def</span> <span class="nf">files</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">File</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of files associated with a phenopacket</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[File]: List of files associated with a phenopacket</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">files</span>

    <span class="k">def</span> <span class="nf">vcf_file_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">File</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the genome assembly and VCF file name from a phenopacket.</span>

<span class="sd">        Args:</span>
<span class="sd">            phenopacket_path (Path): The path to the phenopacket file.</span>
<span class="sd">            vcf_dir (Path): The directory path where the VCF file is stored.</span>

<span class="sd">        Returns:</span>
<span class="sd">            File: The VCF file with updated URI pointing to the specified directory.</span>

<span class="sd">        Raises:</span>
<span class="sd">            IncorrectFileFormatError: If the provided file is not in .vcf or .vcf.gz format.</span>
<span class="sd">            IncompatibleGenomeAssemblyError: If the genome assembly of the VCF file is not compatible.</span>

<span class="sd">        Note:</span>
<span class="sd">            This function searches for a VCF file within the provided list of files, validates its format,</span>
<span class="sd">            and checks if the genome assembly is compatible. If the conditions are met, it updates the</span>
<span class="sd">            URI of the VCF file to the specified directory and returns the modified file object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">compatible_genome_assembly</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">]</span>
        <span class="n">vcf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">()</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;vcf&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span>
            <span class="s2">&quot;.vcf.gz&quot;</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="n">IncorrectFileFormatError</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">),</span> <span class="s2">&quot;.vcf or .vcf.gz file&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">compatible_genome_assembly</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">IncompatibleGenomeAssemblyError</span><span class="p">(</span>
                <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">],</span> <span class="n">phenopacket_path</span>
            <span class="p">)</span>
        <span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">vcf_data</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_extract_diagnosed_gene</span><span class="p">(</span>
        <span class="n">genomic_interpretation</span><span class="p">:</span> <span class="n">GenomicInterpretation</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ProbandCausativeGene</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the disease causing genes from the variant descriptor field if not empty,</span>
<span class="sd">        otherwise, retrieves from the gene descriptor from a phenopacket.</span>
<span class="sd">        Args:</span>
<span class="sd">            genomic_interpretation (GenomicInterpretation): A genomic interpretation from a Phenopacket</span>
<span class="sd">        Returns:</span>
<span class="sd">            ProbandCausativeGene: The disease causing gene</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">ByteSize</span><span class="p">()</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">ProbandCausativeGene</span><span class="p">(</span>
                <span class="n">gene_symbol</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">symbol</span><span class="p">,</span>
                <span class="n">gene_identifier</span><span class="o">=</span><span class="n">genomic_interpretation</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">value_id</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span> <span class="nf">diagnosed_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeGene</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the disease causing genes from a phenopacket.</span>
<span class="sd">        Returns:</span>
<span class="sd">            List[ProbandCausativeGene]: List of causative genes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_diagnosed_gene</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
                <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">}</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">genes</span>

    <span class="k">def</span> <span class="nf">diagnosed_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">GenomicVariant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve a list of all known causative variants from a phenopacket.</span>
<span class="sd">        Returns:</span>
<span class="sd">            List[GenomicVariant]: List of causative variants</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
                <span class="n">variant</span> <span class="o">=</span> <span class="n">GenomicVariant</span><span class="p">(</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">pos</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">ref</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                    <span class="n">alt</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.__init__" class="doc doc-heading">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">phenopacket_contents</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Initialise PhenopacketUtil</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>phenopacket_contents</code></td>
          <td>
                <code><span title="typing.Union">Union</span>[<span title="phenopackets.Phenopacket">Phenopacket</span>, <span title="phenopackets.Family">Family</span>]</code>
          </td>
          <td><p>Phenopacket or Family object</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_contents</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialise PhenopacketUtil</span>

<span class="sd">    Args:</span>
<span class="sd">        phenopacket_contents (Union[Phenopacket, Family]): Phenopacket or Family object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span> <span class="o">=</span> <span class="n">phenopacket_contents</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.causative_variants" class="doc doc-heading">
<code class="highlight language-python"><span class="n">causative_variants</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of causative variants listed in a Phenopacket</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.ProbandCausativeVariant" href="#src.pheval.utils.phenopacket_utils.ProbandCausativeVariant">ProbandCausativeVariant</a>]</code>
          </td>
          <td><p>List[ProbandCausativeVariant]: List of proband causative variants</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">causative_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeVariant</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of causative variants listed in a Phenopacket</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[ProbandCausativeVariant]: List of proband causative variants</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">all_variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span>
            <span class="n">genotype</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">allelic_state</span>
            <span class="n">variant_data</span> <span class="o">=</span> <span class="n">ProbandCausativeVariant</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="n">vcf_record</span><span class="o">.</span><span class="n">genome_assembly</span><span class="p">,</span>
                <span class="n">GenomicVariant</span><span class="p">(</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                    <span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">genotype</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
                <span class="n">vcf_record</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">all_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant_data</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">all_variants</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagnosed_genes</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve the disease causing genes from a phenopacket.</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.ProbandCausativeGene" href="#src.pheval.utils.phenopacket_utils.ProbandCausativeGene">ProbandCausativeGene</a>]</code>
          </td>
          <td><p>List[ProbandCausativeGene]: List of causative genes</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagnosed_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandCausativeGene</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve the disease causing genes from a phenopacket.</span>
<span class="sd">    Returns:</span>
<span class="sd">        List[ProbandCausativeGene]: List of causative genes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_diagnosed_gene</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">:</span> <span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">}</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">genes</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnosed_variants" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagnosed_variants</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of all known causative variants from a phenopacket.</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.GenomicVariant" href="#src.pheval.utils.phenopacket_utils.GenomicVariant">GenomicVariant</a>]</code>
          </td>
          <td><p>List[GenomicVariant]: List of causative variants</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagnosed_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">GenomicVariant</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of all known causative variants from a phenopacket.</span>
<span class="sd">    Returns:</span>
<span class="sd">        List[GenomicVariant]: List of causative variants</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">pheno_interpretation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">interpretations</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pheno_interpretation</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">diagnosis</span><span class="o">.</span><span class="n">genomic_interpretations</span><span class="p">:</span>
            <span class="n">variant</span> <span class="o">=</span> <span class="n">GenomicVariant</span><span class="p">(</span>
                <span class="n">chrom</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span>
                <span class="n">pos</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span>
                <span class="n">ref</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
                <span class="n">alt</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">variant_interpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">alt</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diagnoses" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagnoses</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a unique list of disease diagnoses associated with the proband from a Phenopacket</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.ProbandDisease" href="#src.pheval.utils.phenopacket_utils.ProbandDisease">ProbandDisease</a>]</code>
          </td>
          <td><p>List[ProbandDisease]: List of diagnosed diseases</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagnoses</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">ProbandDisease</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a unique list of disease diagnoses associated with the proband from a Phenopacket</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[ProbandDisease]: List of diagnosed diseases</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_interpretations</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_diagnosis_from_disease</span><span class="p">()))</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.diseases" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diseases</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of Diseases associated with the proband</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.Disease">Disease</span>]</code>
          </td>
          <td><p>List[Disease]: List of diseases</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diseases</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Disease</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of Diseases associated with the proband</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[Disease]: List of diseases</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">diseases</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">diseases</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.files" class="doc doc-heading">
<code class="highlight language-python"><span class="n">files</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of files associated with a phenopacket</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.File">File</span>]</code>
          </td>
          <td><p>List[File]: List of files associated with a phenopacket</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">files</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">File</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of files associated with a phenopacket</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[File]: List of files associated with a phenopacket</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">files</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.interpretations" class="doc doc-heading">
<code class="highlight language-python"><span class="n">interpretations</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of interpretations from a Phenopacket</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.Interpretation">Interpretation</span>]</code>
          </td>
          <td><p>List[Interpretation]: List of interpretations</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">interpretations</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Interpretation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of interpretations from a Phenopacket</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[Interpretation]: List of interpretations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">interpretations</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">interpretations</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.negated_phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">negated_phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of all negated HPO terms</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.PhenotypicFeature">PhenotypicFeature</span>]</code>
          </td>
          <td><p>List[PhenotypicFeature]: List of negated HPO terms</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">negated_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of all negated HPO terms</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[PhenotypicFeature]: List of negated HPO terms</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">negated_phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
            <span class="n">negated_phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">negated_phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.observed_phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">observed_phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of all observed HPO terms</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.PhenotypicFeature">PhenotypicFeature</span>]</code>
          </td>
          <td><p>List[PhenotypicFeature]: List of observed HPO terms</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">observed_phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of all observed HPO terms</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[PhenotypicFeature]: List of observed HPO terms</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenotypic_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_phenotypic_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotypic_features</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">all_phenotypic_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">excluded</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">phenotypic_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.phenotypic_features" class="doc doc-heading">
<code class="highlight language-python"><span class="n">phenotypic_features</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve a list of all HPO terms</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="phenopackets.PhenotypicFeature">PhenotypicFeature</span>]</code>
          </td>
          <td><p>List[PhenotypicFeature]: List of HPO terms</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">phenotypic_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">PhenotypicFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve a list of all HPO terms</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[PhenotypicFeature]: List of HPO terms</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">phenotypic_features</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">phenotypic_features</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.sample_id" class="doc doc-heading">
<code class="highlight language-python"><span class="n">sample_id</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve the sample ID from a Phenopacket or proband of a Family</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>str</code></td>          <td>
                <code>str</code>
          </td>
          <td><p>Sample ID</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sample_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve the sample ID from a Phenopacket or proband of a Family</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Sample ID</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="p">,</span> <span class="s2">&quot;proband&quot;</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">proband</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenopacket_contents</span><span class="o">.</span><span class="n">subject</span><span class="o">.</span><span class="n">id</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="src.pheval.utils.phenopacket_utils.PhenopacketUtil.vcf_file_data" class="doc doc-heading">
<code class="highlight language-python"><span class="n">vcf_file_data</span><span class="p">(</span><span class="n">phenopacket_path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Retrieve the genome assembly and VCF file name from a phenopacket.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>phenopacket_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td><p>The path to the phenopacket file.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>vcf_dir</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td><p>The directory path where the VCF file is stored.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>File</code></td>          <td>
                <code><span title="phenopackets.File">File</span></code>
          </td>
          <td><p>The VCF file with updated URI pointing to the specified directory.</p></td>
        </tr>
    </tbody>
  </table>

  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="pheval.prepare.custom_exceptions.IncorrectFileFormatError">IncorrectFileFormatError</span></code>
          </td>
          <td><p>If the provided file is not in .vcf or .vcf.gz format.</p></td>
        </tr>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError" href="#src.pheval.utils.phenopacket_utils.IncompatibleGenomeAssemblyError">IncompatibleGenomeAssemblyError</a></code>
          </td>
          <td><p>If the genome assembly of the VCF file is not compatible.</p></td>
        </tr>
    </tbody>
  </table>

<details class="note">
  <summary>Note</summary>
  <p>This function searches for a VCF file within the provided list of files, validates its format,
and checks if the genome assembly is compatible. If the conditions are met, it updates the
URI of the VCF file to the specified directory and returns the modified file object.</p>
</details>
      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">vcf_file_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenopacket_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">vcf_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">File</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieve the genome assembly and VCF file name from a phenopacket.</span>

<span class="sd">    Args:</span>
<span class="sd">        phenopacket_path (Path): The path to the phenopacket file.</span>
<span class="sd">        vcf_dir (Path): The directory path where the VCF file is stored.</span>

<span class="sd">    Returns:</span>
<span class="sd">        File: The VCF file with updated URI pointing to the specified directory.</span>

<span class="sd">    Raises:</span>
<span class="sd">        IncorrectFileFormatError: If the provided file is not in .vcf or .vcf.gz format.</span>
<span class="sd">        IncompatibleGenomeAssemblyError: If the genome assembly of the VCF file is not compatible.</span>

<span class="sd">    Note:</span>
<span class="sd">        This function searches for a VCF file within the provided list of files, validates its format,</span>
<span class="sd">        and checks if the genome assembly is compatible. If the conditions are met, it updates the</span>
<span class="sd">        URI of the VCF file to the specified directory and returns the modified file object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">compatible_genome_assembly</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;GRCh37&quot;</span><span class="p">,</span> <span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">]</span>
    <span class="n">vcf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">()</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;fileFormat&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;vcf&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span>
        <span class="s2">&quot;.vcf.gz&quot;</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="n">IncorrectFileFormatError</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">),</span> <span class="s2">&quot;.vcf or .vcf.gz file&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">compatible_genome_assembly</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">IncompatibleGenomeAssemblyError</span><span class="p">(</span>
            <span class="n">vcf_data</span><span class="o">.</span><span class="n">file_attributes</span><span class="p">[</span><span class="s2">&quot;genomeAssembly&quot;</span><span class="p">],</span> <span class="n">phenopacket_path</span>
        <span class="p">)</span>
    <span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">.</span><span class="n">joinpath</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">vcf_data</span><span class="o">.</span><span class="n">uri</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">vcf_data</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.ProbandCausativeGene" class="doc doc-heading">
        <code>ProbandCausativeGene</code>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

</h2>


  <div class="doc doc-contents ">

  
      <p>Represents a causative gene associated with a proband</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene_symbol</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Symbol representing the gene</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gene_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>The ENSEMBL gene identifier for the result entry</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>While we recommend providing the gene identifier in the ENSEMBL namespace,
any matching format used in Phenopacket interpretations and result output is acceptable
for result matching purposes in the analysis.</p>
</details>

        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">ProbandCausativeGene</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents a causative gene associated with a proband</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_symbol (str): Symbol representing the gene</span>
<span class="sd">        gene_identifier (str): The ENSEMBL gene identifier for the result entry</span>
<span class="sd">    Notes:</span>
<span class="sd">         While we recommend providing the gene identifier in the ENSEMBL namespace,</span>
<span class="sd">         any matching format used in Phenopacket interpretations and result output is acceptable</span>
<span class="sd">         for result matching purposes in the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_identifier</span><span class="p">:</span> <span class="nb">str</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.ProbandCausativeVariant" class="doc doc-heading">
        <code>ProbandCausativeVariant</code>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

</h2>


  <div class="doc doc-contents ">

  
      <p>Represents a causative variant associated with a proband</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>proband_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>ID of the proband</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>assembly</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Genome assembly</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>variant</code></td>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.pheval.utils.phenopacket_utils.GenomicVariant" href="#src.pheval.utils.phenopacket_utils.GenomicVariant">GenomicVariant</a></code>
          </td>
          <td><p>Genomic variant associated with the proband</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>genotype</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Genotype information for the variant</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>info</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Additional information about the variant (default is an empty string)</p></td>
          <td>
                <code>&#39;&#39;</code>
          </td>
        </tr>
    </tbody>
  </table>


        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">ProbandCausativeVariant</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents a causative variant associated with a proband</span>

<span class="sd">    Args:</span>
<span class="sd">        proband_id (str): ID of the proband</span>
<span class="sd">        assembly (str): Genome assembly</span>
<span class="sd">        variant (GenomicVariant): Genomic variant associated with the proband</span>
<span class="sd">        genotype (str): Genotype information for the variant</span>
<span class="sd">        info (str, optional): Additional information about the variant (default is an empty string)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">proband_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">variant</span><span class="p">:</span> <span class="n">GenomicVariant</span>
    <span class="n">genotype</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">info</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="src.pheval.utils.phenopacket_utils.ProbandDisease" class="doc doc-heading">
        <code>ProbandDisease</code>

  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

</h2>


  <div class="doc doc-contents ">

  
      <p>Represents a disease associated with a proband</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>disease_name</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Name of the disease</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>disease_identifier</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>Identifier for the disease result entry in the OMIM namespace</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>While we recommend providing the disease identifier in the OMIM namespace,
any matching format used in Phenopacket interpretations and result output is acceptable
for result matching purposes in the analysis.</p>
</details>

        <details class="quote">
          <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span><span class="p">(</span><span class="n">frozen</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">eq</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">class</span> <span class="nc">ProbandDisease</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents a disease associated with a proband</span>

<span class="sd">    Args:</span>
<span class="sd">        disease_name (str): Name of the disease</span>
<span class="sd">        disease_identifier (str): Identifier for the disease result entry in the OMIM namespace</span>

<span class="sd">    Notes:</span>
<span class="sd">         While we recommend providing the disease identifier in the OMIM namespace,</span>
<span class="sd">         any matching format used in Phenopacket interpretations and result output is acceptable</span>
<span class="sd">         for result matching purposes in the analysis.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">disease_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">disease_identifier</span><span class="p">:</span> <span class="nb">str</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.create_gene_identifier_map" class="doc doc-heading">
<code class="highlight language-python"><span class="n">create_gene_identifier_map</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Create a mapping of gene identifiers to gene symbols using HGNC data.</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>dict</code></td>          <td>
                <code>dict</code>
          </td>
          <td><p>A mapping of gene identifiers to gene symbols.</p></td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>The dictionary structure:
{
    'identifier': 'gene_symbol',
    ...
}</p>
</details>
      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_gene_identifier_map</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a mapping of gene identifiers to gene symbols using HGNC data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A mapping of gene identifiers to gene symbols.</span>

<span class="sd">    Notes:</span>
<span class="sd">        The dictionary structure:</span>
<span class="sd">        {</span>
<span class="sd">            &#39;identifier&#39;: &#39;gene_symbol&#39;,</span>
<span class="sd">            ...</span>
<span class="sd">        }</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hgnc_df</span> <span class="o">=</span> <span class="n">read_hgnc_data</span><span class="p">()</span>
    <span class="n">identifier_map</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">hgnc_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">identifier_map</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
        <span class="n">identifier_map</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
        <span class="n">identifier_map</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
        <span class="n">identifier_map</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;refseq_accession&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">identifier_map</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.create_hgnc_dict" class="doc doc-heading">
<code class="highlight language-python"><span class="n">create_hgnc_dict</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Create a dictionary as a reference for updating gene symbols and identifiers based on HGNC data.</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>defaultdict</code></td>          <td>
                <code><span title="collections.defaultdict">defaultdict</span></code>
          </td>
          <td><p>A dictionary containing gene symbols as keys and their associated gene information.</p></td>
        </tr>
    </tbody>
  </table>

<details class="notes">
  <summary>Notes</summary>
  <p>The dictionary structure:
{
    'gene_symbol': {
        'ensembl_id': str,
        'hgnc_id': str,
        'entrez_id': str,
        'refseq_accession': str,
        'previous_symbol': [str, ...]
    },
    ...
}</p>
</details>
      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_hgnc_dict</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">defaultdict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a dictionary as a reference for updating gene symbols and identifiers based on HGNC data.</span>


<span class="sd">    Returns:</span>
<span class="sd">        defaultdict: A dictionary containing gene symbols as keys and their associated gene information.</span>

<span class="sd">    Notes:</span>
<span class="sd">        The dictionary structure:</span>
<span class="sd">        {</span>
<span class="sd">            &#39;gene_symbol&#39;: {</span>
<span class="sd">                &#39;ensembl_id&#39;: str,</span>
<span class="sd">                &#39;hgnc_id&#39;: str,</span>
<span class="sd">                &#39;entrez_id&#39;: str,</span>
<span class="sd">                &#39;refseq_accession&#39;: str,</span>
<span class="sd">                &#39;previous_symbol&#39;: [str, ...]</span>
<span class="sd">            },</span>
<span class="sd">            ...</span>
<span class="sd">        }</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hgnc_df</span> <span class="o">=</span> <span class="n">read_hgnc_data</span><span class="p">()</span>
    <span class="n">hgnc_data</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">_index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">hgnc_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">previous_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;hgnc_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;entrez_id&quot;</span><span class="p">]</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;refseq_accession&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;refseq_accession&quot;</span><span class="p">]</span>
        <span class="n">previous</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;prev_symbol&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">previous</span><span class="p">:</span>
            <span class="n">previous_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">))</span>
        <span class="n">hgnc_data</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]][</span><span class="s2">&quot;previous_symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">previous_names</span>

    <span class="k">return</span> <span class="n">hgnc_data</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.create_json_message" class="doc doc-heading">
<code class="highlight language-python"><span class="n">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Create a JSON message for writing to a file.</p>
      <ul>
<li>phenopacket (Union[Phenopacket, Family]): The Phenopacket or Family object to convert to JSON.</li>
</ul>
      <ul>
<li>str: A JSON-formatted string representation of the Phenopacket or Family object.</li>
</ul>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a JSON message for writing to a file.</span>

<span class="sd">    Args:</span>
<span class="sd">    - phenopacket (Union[Phenopacket, Family]): The Phenopacket or Family object to convert to JSON.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: A JSON-formatted string representation of the Phenopacket or Family object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">MessageToJson</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.phenopacket_reader" class="doc doc-heading">
<code class="highlight language-python"><span class="n">phenopacket_reader</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Read a Phenopacket file and returns its contents as a Phenopacket or Family object</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>file</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td><p>Path to the Phenopacket file</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Union">Union</span>[<span title="phenopackets.Phenopacket">Phenopacket</span>, <span title="phenopackets.Family">Family</span>]</code>
          </td>
          <td><p>Union[Phenopacket, Family]: Contents of the Phenopacket file as a Phenopacket or Family object</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">phenopacket_reader</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read a Phenopacket file and returns its contents as a Phenopacket or Family object</span>

<span class="sd">    Args:</span>
<span class="sd">        file (Path): Path to the Phenopacket file</span>

<span class="sd">    Returns:</span>
<span class="sd">        Union[Phenopacket, Family]: Contents of the Phenopacket file as a Phenopacket or Family object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="n">phenopacket</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;proband&quot;</span> <span class="ow">in</span> <span class="n">phenopacket</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Parse</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">),</span> <span class="n">Family</span><span class="p">())</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Parse</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">),</span> <span class="n">Phenopacket</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.read_hgnc_data" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read_hgnc_data</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Read HGNC data from a file and return it as a Pandas DataFrame.</p>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
          </td>
          <td><p>pd.DataFrame: DataFrame containing the HGNC data.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_hgnc_data</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read HGNC data from a file and return it as a Pandas DataFrame.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame containing the HGNC data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;utils&quot;</span><span class="p">,</span> <span class="s2">&quot;resources/hgnc_complete_set.txt&quot;</span><span class="p">),</span>
        <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h2 id="src.pheval.utils.phenopacket_utils.write_phenopacket" class="doc doc-heading">
<code class="highlight language-python"><span class="n">write_phenopacket</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Write a Phenopacket or Family object to a file in JSON format.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>phenopacket</code></td>
          <td>
                <code>Phenopacket or Family</code>
          </td>
          <td><p>The Phenopacket or Family object to be written.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>output_file</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td><p>The Path object representing the file to write the Phenopacket data.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td><p>None</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>src/pheval/utils/phenopacket_utils.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">write_phenopacket</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Phenopacket</span><span class="p">,</span> <span class="n">Family</span><span class="p">],</span> <span class="n">output_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write a Phenopacket or Family object to a file in JSON format.</span>

<span class="sd">    Args:</span>
<span class="sd">        phenopacket (Phenopacket or Family): The Phenopacket or Family object to be written.</span>
<span class="sd">        output_file (Path): The Path object representing the file to write the Phenopacket data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phenopacket_json</span> <span class="o">=</span> <span class="n">create_json_message</span><span class="p">(</span><span class="n">phenopacket</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
        <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">phenopacket_json</span><span class="p">)</span>
    <span class="n">outfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../file_utils/" class="md-footer__link md-footer__link--prev" aria-label="Previous: File utils" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              File utils
            </div>
          </div>
        </a>
      
      
        
        <a href="../semsim_utils/" class="md-footer__link md-footer__link--next" aria-label="Next: Semsim utils" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Semsim utils
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../../..", "features": [], "search": "../../../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.5a2dcb6a.min.js"></script>
      
        <script src="https://unpkg.com/mermaid@8.5.1/dist/mermaid.min.js"></script>
      
    
    
  </body>
</html>