
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/pheval/api/pheval/cli/">
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-8.5.10">
    
    
      
        <title>Cli - PhEval</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.975780f9.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.2505c338.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="https://unpkg.com/mermaid@8.5.1/dist/mermaid.css">
    
      <link rel="stylesheet" href="../../../mermaid.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#main" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="PhEval" class="md-header__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PhEval
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Cli
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Night Mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Night Mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
            </label>
          
        
          
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Day Mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Day Mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="PhEval" class="md-nav__button md-logo" aria-label="PhEval" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    PhEval
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/pheval" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../contact/" class="md-nav__link">
        Contact Us
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          API Documentation
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="API Documentation" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          API Documentation
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Cli
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Cli
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#main" class="md-nav__link">
    main
  </a>
  
    <nav class="md-nav" aria-label="main">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pheval" class="md-nav__link">
    pheval
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run" class="md-nav__link">
    # run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update" class="md-nav__link">
    # update
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pheval-utils" class="md-nav__link">
    pheval-utils
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#benchmark" class="md-nav__link">
    # benchmark
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create-spiked-vcfs" class="md-nav__link">
    # create-spiked-vcfs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-plots" class="md-nav__link">
    # generate-plots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-corpus" class="md-nav__link">
    # prepare-corpus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scramble-phenopackets" class="md-nav__link">
    # scramble-phenopackets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-scramble" class="md-nav__link">
    # semsim-scramble
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-to-exomiserdb" class="md-nav__link">
    # semsim-to-exomiserdb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update-phenopackets" class="md-nav__link">
    # update-phenopackets
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pheval_1" class="md-nav__link">
    pheval
  </a>
  
    <nav class="md-nav" aria-label="pheval">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run_1" class="md-nav__link">
    run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update_1" class="md-nav__link">
    update
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pheval-utils_1" class="md-nav__link">
    pheval-utils
  </a>
  
    <nav class="md-nav" aria-label="pheval-utils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#benchmark_1" class="md-nav__link">
    benchmark
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create-spiked-vcfs_1" class="md-nav__link">
    create-spiked-vcfs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-plots_1" class="md-nav__link">
    generate-plots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-corpus_1" class="md-nav__link">
    prepare-corpus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scramble-phenopackets_1" class="md-nav__link">
    scramble-phenopackets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-scramble_1" class="md-nav__link">
    semsim-scramble
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-to-exomiserdb_1" class="md-nav__link">
    semsim-to-exomiserdb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update-phenopackets_1" class="md-nav__link">
    update-phenopackets
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../config_parser/" class="md-nav__link">
        Config parser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../run_metadata/" class="md-nav__link">
        Run metadata
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_4" type="checkbox" id="__nav_4_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_4">
          analyse
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="analyse" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_4">
          <span class="md-nav__icon md-icon"></span>
          analyse
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/benchmark/" class="md-nav__link">
        Benchmark
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/benchmark_db_manager/" class="md-nav__link">
        Benchmark db manager
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/benchmark_output_type/" class="md-nav__link">
        Benchmark output type
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/binary_classification_curves/" class="md-nav__link">
        Binary classification curves
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/binary_classification_stats/" class="md-nav__link">
        Binary classification stats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/generate_plots/" class="md-nav__link">
        Generate plots
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/generate_rank_comparisons/" class="md-nav__link">
        Generate rank comparisons
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/rank_stats/" class="md-nav__link">
        Rank stats
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../analyse/run_data_parser/" class="md-nav__link">
        Run data parser
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_5" type="checkbox" id="__nav_4_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_5">
          implementations
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="implementations" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_5">
          <span class="md-nav__icon md-icon"></span>
          implementations
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../implementations/pheval_class_resolver/" class="md-nav__link">
        Pheval class resolver
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_6" type="checkbox" id="__nav_4_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_6">
          infra
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="infra" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_6">
          <span class="md-nav__icon md-icon"></span>
          infra
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../infra/exomiserdb/" class="md-nav__link">
        Exomiserdb
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_7" type="checkbox" id="__nav_4_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_7">
          post_processing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="post_processing" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_7">
          <span class="md-nav__icon md-icon"></span>
          post_processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../post_processing/mondo_mapping/" class="md-nav__link">
        Mondo mapping
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../post_processing/phenopacket_truth_set/" class="md-nav__link">
        Phenopacket truth set
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../post_processing/post_processing/" class="md-nav__link">
        Post processing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../post_processing/validate_result_format/" class="md-nav__link">
        Validate result format
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_8" type="checkbox" id="__nav_4_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_8">
          prepare
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="prepare" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_8">
          <span class="md-nav__icon md-icon"></span>
          prepare
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/create_noisy_phenopackets/" class="md-nav__link">
        Create noisy phenopackets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/create_spiked_vcf/" class="md-nav__link">
        Create spiked vcf
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/custom_exceptions/" class="md-nav__link">
        Custom exceptions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/prepare_corpus/" class="md-nav__link">
        Prepare corpus
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../prepare/update_phenopacket/" class="md-nav__link">
        Update phenopacket
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_9" type="checkbox" id="__nav_4_9" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_9">
          runners
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="runners" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_9">
          <span class="md-nav__icon md-icon"></span>
          runners
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../runners/runner/" class="md-nav__link">
        Runner
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4_10" type="checkbox" id="__nav_4_10" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4_10">
          utils
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="utils" data-md-level="2">
        <label class="md-nav__title" for="__nav_4_10">
          <span class="md-nav__icon md-icon"></span>
          utils
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/exomiser/" class="md-nav__link">
        Exomiser
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/file_utils/" class="md-nav__link">
        File utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/logger/" class="md-nav__link">
        Logger
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/phenopacket_utils/" class="md-nav__link">
        Phenopacket utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/semsim_utils/" class="md-nav__link">
        Semsim utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/utils/" class="md-nav__link">
        Utils
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Resources for contributors
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Resources for contributors" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Resources for contributors
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../developing_a_pheval_plugin/" class="md-nav__link">
        Developing a PhEval Plugin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../contributing/" class="md-nav__link">
        Contributions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../styleguide/" class="md-nav__link">
        Monarch Style Guide for PhEval
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../CODE_OF_CONDUCT/" class="md-nav__link">
        Contributor Covenant Code of Conduct
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../plugins/" class="md-nav__link">
        Plugins
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../executing_a_benchmark/" class="md-nav__link">
        Executing a Benchmark
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../roadmap/" class="md-nav__link">
        Roadmap
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#main" class="md-nav__link">
    main
  </a>
  
    <nav class="md-nav" aria-label="main">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pheval" class="md-nav__link">
    pheval
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run" class="md-nav__link">
    # run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update" class="md-nav__link">
    # update
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pheval-utils" class="md-nav__link">
    pheval-utils
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#benchmark" class="md-nav__link">
    # benchmark
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create-spiked-vcfs" class="md-nav__link">
    # create-spiked-vcfs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-plots" class="md-nav__link">
    # generate-plots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-corpus" class="md-nav__link">
    # prepare-corpus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scramble-phenopackets" class="md-nav__link">
    # scramble-phenopackets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-scramble" class="md-nav__link">
    # semsim-scramble
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-to-exomiserdb" class="md-nav__link">
    # semsim-to-exomiserdb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update-phenopackets" class="md-nav__link">
    # update-phenopackets
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pheval_1" class="md-nav__link">
    pheval
  </a>
  
    <nav class="md-nav" aria-label="pheval">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run_1" class="md-nav__link">
    run
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update_1" class="md-nav__link">
    update
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pheval-utils_1" class="md-nav__link">
    pheval-utils
  </a>
  
    <nav class="md-nav" aria-label="pheval-utils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#benchmark_1" class="md-nav__link">
    benchmark
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create-spiked-vcfs_1" class="md-nav__link">
    create-spiked-vcfs
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#generate-plots_1" class="md-nav__link">
    generate-plots
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-corpus_1" class="md-nav__link">
    prepare-corpus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scramble-phenopackets_1" class="md-nav__link">
    scramble-phenopackets
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-scramble_1" class="md-nav__link">
    semsim-scramble
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#semsim-to-exomiserdb_1" class="md-nav__link">
    semsim-to-exomiserdb
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#update-phenopackets_1" class="md-nav__link">
    update-phenopackets
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/monarch-initiative/pheval/edit/docs/docs/api/pheval/cli.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>Cli</h1>

<h5 id="main">main</h5>
<p>Main CLI method for PhEval.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main [OPTIONS] COMMAND [ARGS]...
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-v</code>, <code>--verbose</code></td>
<td>integer range (<code>0</code> and above)</td>
<td>N/A</td>
<td><code>0</code></td>
</tr>
<tr>
<td><code>-q</code>, <code>--quiet</code></td>
<td>boolean</td>
<td>N/A</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<p><strong>Subcommands</strong></p>
<ul>
<li><em>pheval</em>: pheval</li>
<li><em>pheval-utils</em>: pheval_utils</li>
</ul>
<h6 id="pheval">pheval</h6>
<p>pheval</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval [OPTIONS] COMMAND [ARGS]...
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<p><strong>Subcommands</strong></p>
<ul>
<li><em>run</em>: PhEval Runner Command Line Interface</li>
<li><em>update</em>: </li>
</ul>
<h6 id="run"># run</h6>
<p>PhEval Runner Command Line Interface
Args:
    input_dir (Path): The input directory (relative path: e.g exomiser-13.11)
    testdata_dir (Path): The input directory (relative path: e.g ./data
    runner (str): Runner implementation (e.g exomiser-13.11)
    tmp_dir (Path): The path of the temporary directory (optional)
    output_dir (Path): The path of the output directory
    config (Path): The path of the configuration file (optional e.g., config.yaml)
    version (str): The version of the tool implementation</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval run [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input-dir</code>, <code>-i</code></td>
<td>Path</td>
<td>The input directory (relative path: e.g exomiser-13.11)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--testdata-dir</code>, <code>-t</code></td>
<td>Path</td>
<td>The input directory (relative path: e.g ./data)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--runner</code>, <code>-r</code></td>
<td>text</td>
<td>Runner implementation (e.g exomiser-13.11)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--tmp-dir</code>, <code>-m</code></td>
<td>Path</td>
<td>The path of the temporary directory (optional)</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>The path of the output directory</td>
<td>_required</td>
</tr>
<tr>
<td><code>--config</code>, <code>-c</code></td>
<td>Path</td>
<td>The path of the configuration file (optional e.g config.yaml)</td>
<td>None</td>
</tr>
<tr>
<td><code>--version</code>, <code>-v</code></td>
<td>text</td>
<td>Version of the tool implementation.</td>
<td>None</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="update"># update</h6>
<p>Download the latest MONDO and HGNC mapping files.</p>
<p>This command fetches the most recent versions of:</p>
<ul>
<li>
<p>The MONDO SSSOM mapping file from the Monarch Initiative</p>
</li>
<li>
<p>The HGNC complete gene set from the HGNC download site</p>
</li>
</ul>
<p>These files are saved to the <code>resources/</code> directory and will
overwrite any existing versions. This ensures that PhEval has
access to the most up-to-date identifier mappings for disease
and gene normalisation.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval update [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="pheval-utils">pheval-utils</h6>
<p>pheval_utils</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils [OPTIONS] COMMAND [ARGS]...
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<p><strong>Subcommands</strong></p>
<ul>
<li><em>benchmark</em>: Benchmark the gene/variant/disease prioritisation performance for runs.</li>
<li><em>create-spiked-vcfs</em>: </li>
<li><em>generate-plots</em>: Generate bar plot from benchmark db.</li>
<li><em>prepare-corpus</em>: </li>
<li><em>scramble-phenopackets</em>: Generate noisy phenopackets from existing ones.</li>
<li><em>semsim-scramble</em>: Scrambles semsim profile multiplying score value by scramble factor</li>
<li><em>semsim-to-exomiserdb</em>: ingests semsim file into exomiser phenotypic database</li>
<li><em>update-phenopackets</em>: Update gene symbols and identifiers for phenopackets.</li>
</ul>
<h6 id="benchmark"># benchmark</h6>
<p>Benchmark the gene/variant/disease prioritisation performance for runs.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils benchmark [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--run-yaml</code>, <code>-r</code></td>
<td>Path</td>
<td>Path to yaml configuration file for benchmarking.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="create-spiked-vcfs"># create-spiked-vcfs</h6>
<p>Create spiked VCF from either a Phenopacket or a Phenopacket directory.</p>
<p>Args:
    phenopacket_path (Path): Path to a single Phenopacket file (optional).
    phenopacket_dir (Path): Path to a directory containing Phenopacket files (optional).
    output_dir (Path): The directory to store the generated spiked VCF file(s).
    hg19_template_vcf (Path): Path to the hg19 template VCF file (optional).
    hg38_template_vcf (Path): Path to the hg38 template VCF file (optional).
    hg19_vcf_dir (Path): Path to the directory containing the hg19 VCF files (optional).
    hg38_vcf_dir (Path): Path to the directory containing the hg38 VCF files (optional).</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils create-spiked-vcfs [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopacket directory for updating. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-template-vcf</code>, <code>-hg19</code></td>
<td>Path</td>
<td>Template hg19 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg19_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-template-vcf</code>, <code>-hg38</code></td>
<td>Path</td>
<td>Template hg38 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg38_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-vcf-dir</code>, <code>-hg19-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg19 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg19_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-vcf-dir</code>, <code>-hg38-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg38 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg38_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-O</code></td>
<td>Path</td>
<td>Path for creation of output directory</td>
<td><code>vcf</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="generate-plots"># generate-plots</h6>
<p>Generate bar plot from benchmark db.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils generate-plots [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--benchmark-db</code>, <code>-b</code></td>
<td>Path</td>
<td>Path to benchmark db output by PhEval benchmark commands.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--run-data</code>, <code>-r</code></td>
<td>Path</td>
<td>Path to yaml configuration file for benchmarking.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="prepare-corpus"># prepare-corpus</h6>
<p>Prepare a corpus of Phenopackets for analysis, optionally checking for complete variant records and updating
gene identifiers.</p>
<p>Args:
    phenopacket_dir (Path): The path to the directory containing Phenopackets.
    variant_analysis (bool): If True, check for complete variant records in the Phenopackets.
    gene_analysis (bool): If True, check for complete gene records in the Phenopackets.
    disease_analysis (bool): If True, check for complete disease records in the Phenopackets.
    gene_identifier (str): Identifier for updating gene identifiers, if applicable.
    hg19_template_vcf (Path): Path to the hg19 template VCF file (optional).
    hg38_template_vcf (Path): Path to the hg38 template VCF file (optional).
    hg19_vcf_dir (Path): Path to the directory containing the hg19 VCF files (optional).
    hg38_vcf_dir (Path): Path to the directory containing the hg38 VCF files (optional).
    output_dir (Path): The directory to save the prepared Phenopackets and, optionally, VCF files.
Notes:
    To spike variants into VCFs for variant-based analysis at least one of hg19_template_vcf, hg38_template_vcf,
    hg19_vcf_dir or hg38_vcf_dir is required.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils prepare-corpus [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-dir</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket corpus directory..</td>
<td>_required</td>
</tr>
<tr>
<td><code>--variant-analysis</code> / <code>--no-variant-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete variant records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--gene-analysis</code> / <code>--no-gene-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete gene records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--disease-analysis</code> / <code>--no-disease-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete disease records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--gene-identifier</code>, <code>-g</code></td>
<td>choice (<code>ensembl_id</code> &#x7C; <code>entrez_id</code> &#x7C; <code>hgnc_id</code>)</td>
<td>Gene identifier to update in phenopacket</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-template-vcf</code>, <code>-hg19</code></td>
<td>Path</td>
<td>Template hg19 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg19_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-template-vcf</code>, <code>-hg38</code></td>
<td>Path</td>
<td>Template hg38 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg38_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-vcf-dir</code>, <code>-hg19-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg19 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg19_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-vcf-dir</code>, <code>-hg38-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg38 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg38_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>Path to output prepared corpus.</td>
<td><code>prepared_corpus</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="scramble-phenopackets"># scramble-phenopackets</h6>
<p>Generate noisy phenopackets from existing ones.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils scramble-phenopackets [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopackets directory. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--scramble-factor</code>, <code>-s</code></td>
<td>float</td>
<td>Scramble factor for randomising phenopacket phenotypic profiles.</td>
<td><code>0.5</code></td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-O</code></td>
<td>Path</td>
<td>Path for creation of output directory</td>
<td><code>noisy_phenopackets</code></td>
</tr>
<tr>
<td><code>--local-ontology-cache</code>, <code>-l</code></td>
<td>Path</td>
<td>Path to the local ontology cache, e.g., path to the hp.obo.</td>
<td>None</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="semsim-scramble"># semsim-scramble</h6>
<p>Scrambles semsim profile multiplying score value by scramble factor
Args:
    input (Path): Path file that points out to the semsim profile
    output (Path): Path file that points out to the output file
    score_column (List[str]): Score column(s) that will be scrambled
    scramble_factor (float): Scramble Magnitude</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils semsim-scramble [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input</code>, <code>-i</code></td>
<td>Path</td>
<td>Path to the semantic similarity profile to be scrambled.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--output</code>, <code>-o</code></td>
<td>Path</td>
<td>Path where the scrambled semsim file will be written.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--score-column</code>, <code>-c</code></td>
<td>choice (<code>jaccard_similarity</code> &#x7C; <code>dice_similarity</code> &#x7C; <code>phenodigm_score</code>)</td>
<td>Score column that will be scrambled</td>
<td>_required</td>
</tr>
<tr>
<td><code>--scramble-factor</code>, <code>-s</code></td>
<td>float</td>
<td>Scramble Magnitude (noise)</td>
<td></td>
</tr>
<tr>
<td>that will be applied to semantic similarity score column (e.g. jaccard similarity).</td>
<td><code>0.5</code></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="semsim-to-exomiserdb"># semsim-to-exomiserdb</h6>
<p>ingests semsim file into exomiser phenotypic database</p>
<p>Args:
    input_file (Path): semsim input file. e.g phenio-plus-hp-mp.0.semsimian.tsv
    object_prefix (str): object prefix. e.g. MP
    subject_prefix (str): subject prefix e.g HP
    db_path (Path): Exomiser Phenotypic Database Folder Path. (e.g. /exomiser_folder/2209_phenotype/2209_phenotype/)</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils semsim-to-exomiserdb [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input-file</code>, <code>-i</code></td>
<td>Path</td>
<td>Semsim input file.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--object-prefix</code></td>
<td>text</td>
<td>Object Prefix. e.g. MP</td>
<td>_required</td>
</tr>
<tr>
<td><code>--subject-prefix</code></td>
<td>text</td>
<td>Subject Prefix. e.g. HP</td>
<td>_required</td>
</tr>
<tr>
<td><code>--db-path</code>, <code>-d</code></td>
<td>Path</td>
<td>Exomiser Phenotypic Database Folder Path.</td>
<td></td>
</tr>
<tr>
<td>(e.g. /exomiser_folder/2209_phenotype/2209_phenotype/).</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>This is the path where the phenotypic database folder will be written out.</td>
<td>_required</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="update-phenopackets"># update-phenopackets</h6>
<p>Update gene symbols and identifiers for phenopackets.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>main pheval-utils update-phenopackets [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopacket directory for updating. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>Path to write phenopacket.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--gene-identifier</code>, <code>-g</code></td>
<td>choice (<code>ensembl_id</code> &#x7C; <code>entrez_id</code> &#x7C; <code>hgnc_id</code>)</td>
<td>Gene identifier to add to phenopacket</td>
<td><code>ensembl_id</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h5 id="pheval_1">pheval</h5>
<p>pheval</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval [OPTIONS] COMMAND [ARGS]...
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<p><strong>Subcommands</strong></p>
<ul>
<li><em>run</em>: PhEval Runner Command Line Interface</li>
<li><em>update</em>: </li>
</ul>
<h6 id="run_1">run</h6>
<p>PhEval Runner Command Line Interface
Args:
    input_dir (Path): The input directory (relative path: e.g exomiser-13.11)
    testdata_dir (Path): The input directory (relative path: e.g ./data
    runner (str): Runner implementation (e.g exomiser-13.11)
    tmp_dir (Path): The path of the temporary directory (optional)
    output_dir (Path): The path of the output directory
    config (Path): The path of the configuration file (optional e.g., config.yaml)
    version (str): The version of the tool implementation</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval run [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input-dir</code>, <code>-i</code></td>
<td>Path</td>
<td>The input directory (relative path: e.g exomiser-13.11)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--testdata-dir</code>, <code>-t</code></td>
<td>Path</td>
<td>The input directory (relative path: e.g ./data)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--runner</code>, <code>-r</code></td>
<td>text</td>
<td>Runner implementation (e.g exomiser-13.11)</td>
<td>_required</td>
</tr>
<tr>
<td><code>--tmp-dir</code>, <code>-m</code></td>
<td>Path</td>
<td>The path of the temporary directory (optional)</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>The path of the output directory</td>
<td>_required</td>
</tr>
<tr>
<td><code>--config</code>, <code>-c</code></td>
<td>Path</td>
<td>The path of the configuration file (optional e.g config.yaml)</td>
<td>None</td>
</tr>
<tr>
<td><code>--version</code>, <code>-v</code></td>
<td>text</td>
<td>Version of the tool implementation.</td>
<td>None</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="update_1">update</h6>
<p>Download the latest MONDO and HGNC mapping files.</p>
<p>This command fetches the most recent versions of:</p>
<ul>
<li>
<p>The MONDO SSSOM mapping file from the Monarch Initiative</p>
</li>
<li>
<p>The HGNC complete gene set from the HGNC download site</p>
</li>
</ul>
<p>These files are saved to the <code>resources/</code> directory and will
overwrite any existing versions. This ensures that PhEval has
access to the most up-to-date identifier mappings for disease
and gene normalisation.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval update [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h5 id="pheval-utils_1">pheval-utils</h5>
<p>pheval_utils</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils [OPTIONS] COMMAND [ARGS]...
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<p><strong>Subcommands</strong></p>
<ul>
<li><em>benchmark</em>: Benchmark the gene/variant/disease prioritisation performance for runs.</li>
<li><em>create-spiked-vcfs</em>: </li>
<li><em>generate-plots</em>: Generate bar plot from benchmark db.</li>
<li><em>prepare-corpus</em>: </li>
<li><em>scramble-phenopackets</em>: Generate noisy phenopackets from existing ones.</li>
<li><em>semsim-scramble</em>: Scrambles semsim profile multiplying score value by scramble factor</li>
<li><em>semsim-to-exomiserdb</em>: ingests semsim file into exomiser phenotypic database</li>
<li><em>update-phenopackets</em>: Update gene symbols and identifiers for phenopackets.</li>
</ul>
<h6 id="benchmark_1">benchmark</h6>
<p>Benchmark the gene/variant/disease prioritisation performance for runs.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils benchmark [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--run-yaml</code>, <code>-r</code></td>
<td>Path</td>
<td>Path to yaml configuration file for benchmarking.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="create-spiked-vcfs_1">create-spiked-vcfs</h6>
<p>Create spiked VCF from either a Phenopacket or a Phenopacket directory.</p>
<p>Args:
    phenopacket_path (Path): Path to a single Phenopacket file (optional).
    phenopacket_dir (Path): Path to a directory containing Phenopacket files (optional).
    output_dir (Path): The directory to store the generated spiked VCF file(s).
    hg19_template_vcf (Path): Path to the hg19 template VCF file (optional).
    hg38_template_vcf (Path): Path to the hg38 template VCF file (optional).
    hg19_vcf_dir (Path): Path to the directory containing the hg19 VCF files (optional).
    hg38_vcf_dir (Path): Path to the directory containing the hg38 VCF files (optional).</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils create-spiked-vcfs [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopacket directory for updating. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-template-vcf</code>, <code>-hg19</code></td>
<td>Path</td>
<td>Template hg19 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg19_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-template-vcf</code>, <code>-hg38</code></td>
<td>Path</td>
<td>Template hg38 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg38_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-vcf-dir</code>, <code>-hg19-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg19 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg19_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-vcf-dir</code>, <code>-hg38-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg38 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg38_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-O</code></td>
<td>Path</td>
<td>Path for creation of output directory</td>
<td><code>vcf</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="generate-plots_1">generate-plots</h6>
<p>Generate bar plot from benchmark db.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils generate-plots [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--benchmark-db</code>, <code>-b</code></td>
<td>Path</td>
<td>Path to benchmark db output by PhEval benchmark commands.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--run-data</code>, <code>-r</code></td>
<td>Path</td>
<td>Path to yaml configuration file for benchmarking.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="prepare-corpus_1">prepare-corpus</h6>
<p>Prepare a corpus of Phenopackets for analysis, optionally checking for complete variant records and updating
gene identifiers.</p>
<p>Args:
    phenopacket_dir (Path): The path to the directory containing Phenopackets.
    variant_analysis (bool): If True, check for complete variant records in the Phenopackets.
    gene_analysis (bool): If True, check for complete gene records in the Phenopackets.
    disease_analysis (bool): If True, check for complete disease records in the Phenopackets.
    gene_identifier (str): Identifier for updating gene identifiers, if applicable.
    hg19_template_vcf (Path): Path to the hg19 template VCF file (optional).
    hg38_template_vcf (Path): Path to the hg38 template VCF file (optional).
    hg19_vcf_dir (Path): Path to the directory containing the hg19 VCF files (optional).
    hg38_vcf_dir (Path): Path to the directory containing the hg38 VCF files (optional).
    output_dir (Path): The directory to save the prepared Phenopackets and, optionally, VCF files.
Notes:
    To spike variants into VCFs for variant-based analysis at least one of hg19_template_vcf, hg38_template_vcf,
    hg19_vcf_dir or hg38_vcf_dir is required.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils prepare-corpus [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-dir</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket corpus directory..</td>
<td>_required</td>
</tr>
<tr>
<td><code>--variant-analysis</code> / <code>--no-variant-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete variant records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--gene-analysis</code> / <code>--no-gene-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete gene records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--disease-analysis</code> / <code>--no-disease-analysis</code></td>
<td>boolean</td>
<td>Specify whether to check for complete disease records in the phenopackets.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>--gene-identifier</code>, <code>-g</code></td>
<td>choice (<code>ensembl_id</code> &#x7C; <code>entrez_id</code> &#x7C; <code>hgnc_id</code>)</td>
<td>Gene identifier to update in phenopacket</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-template-vcf</code>, <code>-hg19</code></td>
<td>Path</td>
<td>Template hg19 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg19_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-template-vcf</code>, <code>-hg38</code></td>
<td>Path</td>
<td>Template hg38 VCF file NOTE: This argument is mutually exclusive with  arguments: [hg38_vcf_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg19-vcf-dir</code>, <code>-hg19-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg19 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg19_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--hg38-vcf-dir</code>, <code>-hg38-dir</code></td>
<td>Path</td>
<td>Path to directory containing hg38 VCF templates. NOTE: This argument is mutually exclusive with  arguments: [hg38_template_vcf].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>Path to output prepared corpus.</td>
<td><code>prepared_corpus</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="scramble-phenopackets_1">scramble-phenopackets</h6>
<p>Generate noisy phenopackets from existing ones.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils scramble-phenopackets [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopackets directory. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--scramble-factor</code>, <code>-s</code></td>
<td>float</td>
<td>Scramble factor for randomising phenopacket phenotypic profiles.</td>
<td><code>0.5</code></td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-O</code></td>
<td>Path</td>
<td>Path for creation of output directory</td>
<td><code>noisy_phenopackets</code></td>
</tr>
<tr>
<td><code>--local-ontology-cache</code>, <code>-l</code></td>
<td>Path</td>
<td>Path to the local ontology cache, e.g., path to the hp.obo.</td>
<td>None</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="semsim-scramble_1">semsim-scramble</h6>
<p>Scrambles semsim profile multiplying score value by scramble factor
Args:
    input (Path): Path file that points out to the semsim profile
    output (Path): Path file that points out to the output file
    score_column (List[str]): Score column(s) that will be scrambled
    scramble_factor (float): Scramble Magnitude</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils semsim-scramble [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input</code>, <code>-i</code></td>
<td>Path</td>
<td>Path to the semantic similarity profile to be scrambled.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--output</code>, <code>-o</code></td>
<td>Path</td>
<td>Path where the scrambled semsim file will be written.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--score-column</code>, <code>-c</code></td>
<td>choice (<code>jaccard_similarity</code> &#x7C; <code>dice_similarity</code> &#x7C; <code>phenodigm_score</code>)</td>
<td>Score column that will be scrambled</td>
<td>_required</td>
</tr>
<tr>
<td><code>--scramble-factor</code>, <code>-s</code></td>
<td>float</td>
<td>Scramble Magnitude (noise)</td>
<td></td>
</tr>
<tr>
<td>that will be applied to semantic similarity score column (e.g. jaccard similarity).</td>
<td><code>0.5</code></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="semsim-to-exomiserdb_1">semsim-to-exomiserdb</h6>
<p>ingests semsim file into exomiser phenotypic database</p>
<p>Args:
    input_file (Path): semsim input file. e.g phenio-plus-hp-mp.0.semsimian.tsv
    object_prefix (str): object prefix. e.g. MP
    subject_prefix (str): subject prefix e.g HP
    db_path (Path): Exomiser Phenotypic Database Folder Path. (e.g. /exomiser_folder/2209_phenotype/2209_phenotype/)</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils semsim-to-exomiserdb [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input-file</code>, <code>-i</code></td>
<td>Path</td>
<td>Semsim input file.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--object-prefix</code></td>
<td>text</td>
<td>Object Prefix. e.g. MP</td>
<td>_required</td>
</tr>
<tr>
<td><code>--subject-prefix</code></td>
<td>text</td>
<td>Subject Prefix. e.g. HP</td>
<td>_required</td>
</tr>
<tr>
<td><code>--db-path</code>, <code>-d</code></td>
<td>Path</td>
<td>Exomiser Phenotypic Database Folder Path.</td>
<td></td>
</tr>
<tr>
<td>(e.g. /exomiser_folder/2209_phenotype/2209_phenotype/).</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>This is the path where the phenotypic database folder will be written out.</td>
<td>_required</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>
<h6 id="update-phenopackets_1">update-phenopackets</h6>
<p>Update gene symbols and identifiers for phenopackets.</p>
<p><strong>Usage:</strong></p>
<div class="highlight"><pre><span></span><code>pheval-utils update-phenopackets [OPTIONS]
</code></pre></div>
<p><strong>Options:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--phenopacket-path</code>, <code>-p</code></td>
<td>Path</td>
<td>Path to phenopacket. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_dir].</td>
<td>None</td>
</tr>
<tr>
<td><code>--phenopacket-dir</code>, <code>-P</code></td>
<td>Path</td>
<td>Path to phenopacket directory for updating. NOTE: This argument is mutually exclusive with  arguments: [phenopacket_path].</td>
<td>None</td>
</tr>
<tr>
<td><code>--output-dir</code>, <code>-o</code></td>
<td>Path</td>
<td>Path to write phenopacket.</td>
<td>_required</td>
</tr>
<tr>
<td><code>--gene-identifier</code>, <code>-g</code></td>
<td>choice (<code>ensembl_id</code> &#x7C; <code>entrez_id</code> &#x7C; <code>hgnc_id</code>)</td>
<td>Gene identifier to add to phenopacket</td>
<td><code>ensembl_id</code></td>
</tr>
<tr>
<td><code>--help</code></td>
<td>boolean</td>
<td>Show this message and exit.</td>
<td><code>False</code></td>
</tr>
</tbody>
</table>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../../../contact/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Contact Us" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Contact Us
            </div>
          </div>
        </a>
      
      
        
        <a href="../config_parser/" class="md-footer__link md-footer__link--next" aria-label="Next: Config parser" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Config parser
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.5a2dcb6a.min.js"></script>
      
        <script src="https://unpkg.com/mermaid@8.5.1/dist/mermaid.min.js"></script>
      
    
    
  </body>
</html>